diff --git a/packages/backend/assets/apple-touch-icon-original.png b/packages/backend/assets/apple-touch-icon-original.png
new file mode 100644
index 0000000000..06ad3f1bb4
Binary files /dev/null and b/packages/backend/assets/apple-touch-icon-original.png differ
diff --git a/packages/backend/assets/apple-touch-icon.png b/packages/backend/assets/apple-touch-icon.png
index 06ad3f1bb4..3a2a269cd6 100644
Binary files a/packages/backend/assets/apple-touch-icon.png and b/packages/backend/assets/apple-touch-icon.png differ
diff --git a/packages/backend/assets/favicon-original.ico b/packages/backend/assets/favicon-original.ico
new file mode 100644
index 0000000000..9be1ff6295
Binary files /dev/null and b/packages/backend/assets/favicon-original.ico differ
diff --git a/packages/backend/assets/favicon-original.png b/packages/backend/assets/favicon-original.png
new file mode 100644
index 0000000000..b4eb18a5cb
Binary files /dev/null and b/packages/backend/assets/favicon-original.png differ
diff --git a/packages/backend/assets/favicon.ico b/packages/backend/assets/favicon.ico
index 9be1ff6295..ac09c3b2c7 100644
Binary files a/packages/backend/assets/favicon.ico and b/packages/backend/assets/favicon.ico differ
diff --git a/packages/backend/assets/favicon.png b/packages/backend/assets/favicon.png
index b4eb18a5cb..b4240ef6b7 100644
Binary files a/packages/backend/assets/favicon.png and b/packages/backend/assets/favicon.png differ
diff --git a/packages/backend/assets/splash-original.png b/packages/backend/assets/splash-original.png
new file mode 100644
index 0000000000..3430e6efe7
Binary files /dev/null and b/packages/backend/assets/splash-original.png differ
diff --git a/packages/backend/assets/splash.png b/packages/backend/assets/splash.png
index 3430e6efe7..ed35cfdf52 100644
Binary files a/packages/backend/assets/splash.png and b/packages/backend/assets/splash.png differ
diff --git a/packages/backend/migration/1700823392948-user-pinned-gallery.js b/packages/backend/migration/1700823392948-user-pinned-gallery.js
new file mode 100644
index 0000000000..85d185be87
--- /dev/null
+++ b/packages/backend/migration/1700823392948-user-pinned-gallery.js
@@ -0,0 +1,13 @@
+export class UserPinnedGallery1700823392948 {
+    name = 'UserPinnedGallery1700823392948'
+
+    async up(queryRunner) {
+        await queryRunner.query(`ALTER TABLE "user_profile" ADD "pinnedGalleryPostId" character varying(32)`);
+        await queryRunner.query(`ALTER TABLE "user_profile" ADD CONSTRAINT "FK_1e9b0b6b3b5f5e2b2f7d7c6f0e9" FOREIGN KEY ("pinnedGalleryPostId") REFERENCES "gallery_post"("id") ON DELETE SET NULL ON UPDATE NO ACTION`);
+    }
+
+    async down(queryRunner) {
+        await queryRunner.query(`ALTER TABLE "user_profile" DROP CONSTRAINT "FK_1e9b0b6b3b5f5e2b2f7d7c6f0e9"`);
+        await queryRunner.query(`ALTER TABLE "user_profile" DROP COLUMN "pinnedGalleryPostId"`);
+    }
+}
diff --git a/packages/backend/migration/1709744820502-bannedMediaDomains.js b/packages/backend/migration/1709744820502-bannedMediaDomains.js
new file mode 100644
index 0000000000..9c4ca89a1c
--- /dev/null
+++ b/packages/backend/migration/1709744820502-bannedMediaDomains.js
@@ -0,0 +1,18 @@
+/*
+ * SPDX-FileCopyrightText: syuilo and misskey-project
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+export class bannedEmailDomains1709744820502 {
+    constructor() {
+        this.name = 'bannedEmailDomains1709744820502';
+    }
+
+    async up(queryRunner) {
+        await queryRunner.query(`ALTER TABLE "meta" ADD "bannedMediaDomains" character varying(1024) array NOT NULL DEFAULT '{}'`);
+    }
+
+    async down(queryRunner) {
+        await queryRunner.query(`ALTER TABLE "meta" DROP COLUMN "bannedMediaDomains"`);
+    }
+}
diff --git a/packages/backend/src/core/entities/UserEntityService.ts b/packages/backend/src/core/entities/UserEntityService.ts
index fbd3892dd4..8f700ca7b6 100644
--- a/packages/backend/src/core/entities/UserEntityService.ts
+++ b/packages/backend/src/core/entities/UserEntityService.ts
@@ -24,9 +24,13 @@ import {
 	passwordSchema,
 } from '@/models/User.js';
 import type {
+	AnnouncementReadsRepository,
+	AnnouncementsRepository,
 	BlockingsRepository,
+	DriveFilesRepository,
 	FollowingsRepository,
 	FollowRequestsRepository,
+	GalleryPostsRepository,
 	MiFollowing,
 	MiUserNotePining,
 	MiUserProfile,
@@ -47,10 +51,12 @@ import { IdService } from '@/core/IdService.js';
 import type { AnnouncementService } from '@/core/AnnouncementService.js';
 import type { CustomEmojiService } from '@/core/CustomEmojiService.js';
 import { AvatarDecorationService } from '@/core/AvatarDecorationService.js';
+import { isNotNull } from '@/misc/is-not-null.js';
+import { DriveFileEntityService } from './DriveFileEntityService.js';
 import type { OnModuleInit } from '@nestjs/common';
 import type { NoteEntityService } from './NoteEntityService.js';
-import type { DriveFileEntityService } from './DriveFileEntityService.js';
 import type { PageEntityService } from './PageEntityService.js';
+import type { GalleryPostEntityService } from './GalleryPostEntityService.js';
 
 const Ajv = _Ajv.default;
 const ajv = new Ajv();
@@ -87,6 +93,7 @@ export class UserEntityService implements OnModuleInit {
 	private apPersonService: ApPersonService;
 	private noteEntityService: NoteEntityService;
 	private pageEntityService: PageEntityService;
+	private galleryPostEntityService: GalleryPostEntityService;
 	private customEmojiService: CustomEmojiService;
 	private announcementService: AnnouncementService;
 	private roleService: RoleService;
@@ -133,6 +140,15 @@ export class UserEntityService implements OnModuleInit {
 		@Inject(DI.userProfilesRepository)
 		private userProfilesRepository: UserProfilesRepository,
 
+		@Inject(DI.announcementReadsRepository)
+		private announcementReadsRepository: AnnouncementReadsRepository,
+
+		@Inject(DI.announcementsRepository)
+		private announcementsRepository: AnnouncementsRepository,
+
+		@Inject(DI.galleryPostsRepository)
+		private galleryPostsRepository: GalleryPostsRepository,
+
 		@Inject(DI.userMemosRepository)
 		private userMemosRepository: UserMemoRepository,
 	) {
@@ -142,6 +158,7 @@ export class UserEntityService implements OnModuleInit {
 		this.apPersonService = this.moduleRef.get('ApPersonService');
 		this.noteEntityService = this.moduleRef.get('NoteEntityService');
 		this.pageEntityService = this.moduleRef.get('PageEntityService');
+		this.galleryPostEntityService = this.moduleRef.get('GalleryPostEntityService');
 		this.customEmojiService = this.moduleRef.get('CustomEmojiService');
 		this.announcementService = this.moduleRef.get('AnnouncementService');
 		this.roleService = this.moduleRef.get('RoleService');
@@ -467,6 +484,7 @@ export class UserEntityService implements OnModuleInit {
 
 		const isModerator = isMe && isDetailed ? this.roleService.isModerator(user) : null;
 		const isAdmin = isMe && isDetailed ? this.roleService.isAdministrator(user) : null;
+		const pinnedGalleryPost = isDetailed && profile?.pinnedGalleryPostId ? profile.pinnedGalleryPost ?? this.galleryPostsRepository.findOneBy({ id: profile.pinnedGalleryPostId, userId: user.id }) : null;
 		const unreadAnnouncements = isMe && isDetailed ?
 			(await this.announcementService.getUnreadAnnouncements(user)).map((announcement) => ({
 				createdAt: this.idService.parse(announcement.id).date.toISOString(),
@@ -482,7 +500,7 @@ export class UserEntityService implements OnModuleInit {
 			host: user.host,
 			avatarUrl: user.avatarUrl ?? this.getIdenticonUrl(user),
 			avatarBlurhash: user.avatarBlurhash,
-			avatarDecorations: user.avatarDecorations.length > 0 ? this.avatarDecorationService.getAll().then(decorations => user.avatarDecorations.filter(ud => decorations.some(d => d.id === ud.id)).map(ud => ({
+			avatarDecorations: user.avatarDecorations.length > 0 ? Promise.all([this.avatarDecorationService.getAll(), this.roleService.getUserPolicies(user.id)]).then(([decorations, policies]) => user.avatarDecorations.filter(ud => decorations.some(d => d.id === ud.id)).slice(0, policies.avatarDecorationLimit).map(ud => ({
 				id: ud.id,
 				angle: ud.angle || undefined,
 				flipH: ud.flipH || undefined,
@@ -547,6 +565,8 @@ export class UserEntityService implements OnModuleInit {
 				}),
 				pinnedPageId: profile!.pinnedPageId,
 				pinnedPage: profile!.pinnedPageId ? this.pageEntityService.pack(profile!.pinnedPageId, me) : null,
+				pinnedGalleryPostId: profile!.pinnedGalleryPostId,
+				pinnedGalleryPost: pinnedGalleryPost ? Promise.resolve(pinnedGalleryPost).then(src => src?.userId === user.id ? this.galleryPostEntityService.pack(src, me) : null) : null,
 				publicReactions: this.isLocalUser(user) ? profile!.publicReactions : false, // https://github.com/misskey-dev/misskey/issues/12964
 				followersVisibility: profile!.followersVisibility,
 				followingVisibility: profile!.followingVisibility,
@@ -567,9 +587,7 @@ export class UserEntityService implements OnModuleInit {
 			...(isDetailed && (isMe || iAmModerator) ? {
 				twoFactorEnabled: profile!.twoFactorEnabled,
 				usePasswordLessLogin: profile!.usePasswordLessLogin,
-				securityKeys: profile!.twoFactorEnabled
-					? this.userSecurityKeysRepository.countBy({ userId: user.id }).then(result => result >= 1)
-					: false,
+				securityKeys: this.userSecurityKeysRepository.countBy({ userId: user.id }).then(result => result >= 1),
 			} : {}),
 
 			...(isDetailed && isMe ? {
@@ -619,18 +637,16 @@ export class UserEntityService implements OnModuleInit {
 			...(opts.includeSecrets ? {
 				email: profile!.email,
 				emailVerified: profile!.emailVerified,
-				securityKeysList: profile!.twoFactorEnabled
-					? this.userSecurityKeysRepository.find({
-						where: {
-							userId: user.id,
-						},
-						select: {
-							id: true,
-							name: true,
-							lastUsed: true,
-						},
-					})
-					: [],
+				securityKeysList: this.userSecurityKeysRepository.find({
+					where: {
+						userId: user.id,
+					},
+					select: {
+						id: true,
+						name: true,
+						lastUsed: true,
+					},
+				}),
 			} : {}),
 
 			...(relation ? {
diff --git a/packages/backend/src/models/Meta.ts b/packages/backend/src/models/Meta.ts
index ad5e31ad6f..1219b60256 100644
--- a/packages/backend/src/models/Meta.ts
+++ b/packages/backend/src/models/Meta.ts
@@ -569,6 +569,13 @@ export class MiMeta {
 	})
 	public bannedEmailDomains: string[];
 
+	@Column('varchar', {
+		length: 1024,
+		array: true,
+		default: '{}',
+	})
+	public bannedMediaDomains: string[];
+
 	@Column('varchar', {
 		length: 1024, array: true, default: '{ "admin", "administrator", "root", "system", "maintainer", "host", "mod", "moderator", "owner", "superuser", "staff", "auth", "i", "me", "everyone", "all", "mention", "mentions", "example", "user", "users", "account", "accounts", "official", "help", "helps", "support", "supports", "info", "information", "informations", "announce", "announces", "announcement", "announcements", "notice", "notification", "notifications", "dev", "developer", "developers", "tech", "misskey" }',
 	})
diff --git a/packages/backend/src/models/UserProfile.ts b/packages/backend/src/models/UserProfile.ts
index 5544555296..8005eae8fa 100644
--- a/packages/backend/src/models/UserProfile.ts
+++ b/packages/backend/src/models/UserProfile.ts
@@ -9,6 +9,7 @@ import { id } from './util/id.js';
 import { MiUser } from './User.js';
 import { MiPage } from './Page.js';
 import { MiUserList } from './UserList.js';
+import { MiGalleryPost } from './GalleryPost.js';
 
 // TODO: このテーブルで管理している情報すべてレジストリで管理するようにしても良いかも
 //       ただ、「emailVerified が true なユーザーを find する」のようなクエリは書けなくなるからウーン
@@ -221,6 +222,18 @@ export class MiUserProfile {
 	@JoinColumn()
 	public pinnedPage: MiPage | null;
 
+	@Column({
+		...id(),
+		nullable: true,
+	})
+	public pinnedGalleryPostId: MiGalleryPost['id'] | null;
+
+	@OneToOne(type => MiGalleryPost, {
+		onDelete: 'SET NULL',
+	})
+	@JoinColumn()
+	public pinnedGalleryPost: MiGalleryPost | null;
+
 	@Index()
 	@Column('boolean', {
 		default: false, select: false,
diff --git a/packages/backend/src/server/FileServerService.ts b/packages/backend/src/server/FileServerService.ts
index bf0a011699..7fbac43898 100644
--- a/packages/backend/src/server/FileServerService.ts
+++ b/packages/backend/src/server/FileServerService.ts
@@ -11,13 +11,14 @@ import rename from 'rename';
 import sharp from 'sharp';
 import { sharpBmp } from '@misskey-dev/sharp-read-bmp';
 import type { Config } from '@/config.js';
-import type { MiDriveFile, DriveFilesRepository } from '@/models/_.js';
+import type { MiDriveFile, MiMeta, DriveFilesRepository } from '@/models/_.js';
 import { DI } from '@/di-symbols.js';
 import { createTemp } from '@/misc/create-temp.js';
 import { FILE_TYPE_BROWSERSAFE } from '@/const.js';
 import { StatusError } from '@/misc/status-error.js';
 import type Logger from '@/logger.js';
 import { DownloadService } from '@/core/DownloadService.js';
+import { MetaService } from '@/core/MetaService.js';
 import { IImageStreamable, ImageProcessingService, webpDefault } from '@/core/ImageProcessingService.js';
 import { VideoProcessingService } from '@/core/VideoProcessingService.js';
 import { InternalStorageService } from '@/core/InternalStorageService.js';
@@ -43,6 +44,9 @@ export class FileServerService {
 		@Inject(DI.config)
 		private config: Config,
 
+		@Inject(DI.meta)
+		private meta: MiMeta,
+
 		@Inject(DI.driveFilesRepository)
 		private driveFilesRepository: DriveFilesRepository,
 
@@ -294,11 +298,23 @@ export class FileServerService {
 	private async proxyHandler(request: FastifyRequest<{ Params: { url: string; }; Querystring: { url?: string; }; }>, reply: FastifyReply) {
 		const url = 'url' in request.query ? request.query.url : 'https://' + request.params.url;
 
-		if (typeof url !== 'string') {
+		if (!(URL.canParse as (url: unknown) => url is string)(url)) {
 			reply.code(400);
 			return;
 		}
 
+		const parsed = new URL(url);
+
+		for (const domain of this.meta.bannedMediaDomains) {
+			const withProtocol = `https://${domain}`;
+			if (!URL.canParse(withProtocol)) continue;
+			const url = new URL(withProtocol);
+			if (`.${parsed.hostname.toLowerCase()}`.endsWith(`.${url.hostname.toLowerCase()}`)) {
+				reply.code(403);
+				return;
+			}
+		}
+
 		// アバタークロップなど、どうしてもオリジンである必要がある場合
 		const mustOrigin = 'origin' in request.query;
 
diff --git a/packages/backend/src/server/api/endpoints/admin/ad/create.ts b/packages/backend/src/server/api/endpoints/admin/ad/create.ts
index 955154f4fb..595d00938f 100644
--- a/packages/backend/src/server/api/endpoints/admin/ad/create.ts
+++ b/packages/backend/src/server/api/endpoints/admin/ad/create.ts
@@ -14,7 +14,7 @@ export const meta = {
 	tags: ['admin'],
 
 	requireCredential: true,
-	requireModerator: true,
+	requireRolePolicy: 'canManageAds',
 	kind: 'write:admin:ad',
 	res: {
 		type: 'object',
diff --git a/packages/backend/src/server/api/endpoints/admin/ad/delete.ts b/packages/backend/src/server/api/endpoints/admin/ad/delete.ts
index 501e13c6a7..32168ea599 100644
--- a/packages/backend/src/server/api/endpoints/admin/ad/delete.ts
+++ b/packages/backend/src/server/api/endpoints/admin/ad/delete.ts
@@ -14,7 +14,7 @@ export const meta = {
 	tags: ['admin'],
 
 	requireCredential: true,
-	requireModerator: true,
+	requireRolePolicy: 'canManageAds',
 	kind: 'write:admin:ad',
 
 	errors: {
diff --git a/packages/backend/src/server/api/endpoints/admin/ad/list.ts b/packages/backend/src/server/api/endpoints/admin/ad/list.ts
index 6406709cda..c4fe0cf96e 100644
--- a/packages/backend/src/server/api/endpoints/admin/ad/list.ts
+++ b/packages/backend/src/server/api/endpoints/admin/ad/list.ts
@@ -13,7 +13,7 @@ export const meta = {
 	tags: ['admin'],
 
 	requireCredential: true,
-	requireModerator: true,
+	requireRolePolicy: 'canManageAds',
 	kind: 'read:admin:ad',
 	res: {
 		type: 'array',
diff --git a/packages/backend/src/server/api/endpoints/admin/ad/update.ts b/packages/backend/src/server/api/endpoints/admin/ad/update.ts
index 4e3d731aca..d0090b3918 100644
--- a/packages/backend/src/server/api/endpoints/admin/ad/update.ts
+++ b/packages/backend/src/server/api/endpoints/admin/ad/update.ts
@@ -14,7 +14,7 @@ export const meta = {
 	tags: ['admin'],
 
 	requireCredential: true,
-	requireModerator: true,
+	requireRolePolicy: 'canManageAds',
 	kind: 'write:admin:ad',
 
 	errors: {
diff --git a/packages/backend/src/server/api/endpoints/admin/meta.ts b/packages/backend/src/server/api/endpoints/admin/meta.ts
index 912c8defbe..11122afb3a 100644
--- a/packages/backend/src/server/api/endpoints/admin/meta.ts
+++ b/packages/backend/src/server/api/endpoints/admin/meta.ts
@@ -192,6 +192,14 @@ export const meta = {
 					optional: false, nullable: false,
 				},
 			},
+			bannedMediaDomains: {
+				type: 'array',
+				optional: true, nullable: false,
+				items: {
+					type: 'string',
+					optional: false, nullable: false,
+				},
+			},
 			preservedUsernames: {
 				type: 'array',
 				optional: false, nullable: false,
@@ -644,6 +652,7 @@ export default class extends Endpoint<typeof meta, typeof paramDef> { // eslint-
 				enableServerMachineStats: instance.enableServerMachineStats,
 				enableIdenticonGeneration: instance.enableIdenticonGeneration,
 				bannedEmailDomains: instance.bannedEmailDomains,
+				bannedMediaDomains: instance.bannedMediaDomains,
 				policies: { ...DEFAULT_POLICIES, ...instance.policies },
 				manifestJsonOverride: instance.manifestJsonOverride,
 				enableFanoutTimeline: instance.enableFanoutTimeline,
diff --git a/packages/backend/src/server/api/endpoints/admin/update-meta.ts b/packages/backend/src/server/api/endpoints/admin/update-meta.ts
index 38ef0d1de8..c017e86b2d 100644
--- a/packages/backend/src/server/api/endpoints/admin/update-meta.ts
+++ b/packages/backend/src/server/api/endpoints/admin/update-meta.ts
@@ -141,6 +141,7 @@ export const paramDef = {
 		enableIdenticonGeneration: { type: 'boolean' },
 		serverRules: { type: 'array', items: { type: 'string' } },
 		bannedEmailDomains: { type: 'array', items: { type: 'string' } },
+		bannedMediaDomains: { type: 'array', items: { type: 'string' } },
 		preservedUsernames: { type: 'array', items: { type: 'string' } },
 		manifestJsonOverride: { type: 'string' },
 		enableFanoutTimeline: { type: 'boolean' },
@@ -639,6 +640,10 @@ export default class extends Endpoint<typeof meta, typeof paramDef> { // eslint-
 				set.bannedEmailDomains = ps.bannedEmailDomains;
 			}
 
+			if (ps.bannedMediaDomains !== undefined) {
+				set.bannedMediaDomains = ps.bannedMediaDomains;
+			}
+
 			if (ps.urlPreviewEnabled !== undefined) {
 				set.urlPreviewEnabled = ps.urlPreviewEnabled;
 			}
diff --git a/packages/backend/src/server/api/endpoints/i/update.ts b/packages/backend/src/server/api/endpoints/i/update.ts
index 4c72879b73..021a5252eb 100644
--- a/packages/backend/src/server/api/endpoints/i/update.ts
+++ b/packages/backend/src/server/api/endpoints/i/update.ts
@@ -11,7 +11,7 @@ import { JSDOM } from 'jsdom';
 import { extractCustomEmojisFromMfm } from '@/misc/extract-custom-emojis-from-mfm.js';
 import { extractHashtags } from '@/misc/extract-hashtags.js';
 import * as Acct from '@/misc/acct.js';
-import type { UsersRepository, DriveFilesRepository, MiMeta, UserProfilesRepository, PagesRepository } from '@/models/_.js';
+import type { UsersRepository, DriveFilesRepository, MiMeta, UserProfilesRepository, PagesRepository, GalleryPostsRepository } from '@/models/_.js';
 import type { MiLocalUser, MiUser } from '@/models/User.js';
 import { birthdaySchema, descriptionSchema, followedMessageSchema, locationSchema, nameSchema } from '@/models/User.js';
 import type { MiUserProfile } from '@/models/UserProfile.js';
@@ -191,6 +191,7 @@ export const paramDef = {
 		followingVisibility: { type: 'string', enum: ['public', 'followers', 'private'] },
 		followersVisibility: { type: 'string', enum: ['public', 'followers', 'private'] },
 		pinnedPageId: { type: 'string', format: 'misskey:id', nullable: true },
+		pinnedGalleryPostId: { type: 'string', format: 'misskey:id', nullable: true },
 		mutedWords: muteWords,
 		hardMutedWords: muteWords,
 		mutedInstances: { type: 'array', items: {
@@ -249,6 +250,9 @@ export default class extends Endpoint<typeof meta, typeof paramDef> { // eslint-
 		@Inject(DI.pagesRepository)
 		private pagesRepository: PagesRepository,
 
+		@Inject(DI.galleryPostsRepository)
+		private galleryPostsRepository: GalleryPostsRepository,
+
 		private userEntityService: UserEntityService,
 		private driveFileEntityService: DriveFileEntityService,
 		private globalEventService: GlobalEventService,
@@ -417,6 +421,16 @@ export default class extends Endpoint<typeof meta, typeof paramDef> { // eslint-
 				profileUpdates.pinnedPageId = null;
 			}
 
+			if (ps.pinnedGalleryPostId) {
+				const galleryPost = await this.galleryPostsRepository.findOneBy({ id: ps.pinnedGalleryPostId, userId: user.id });
+
+				if (galleryPost == null) throw new ApiError(meta.errors.noSuchPage); // TODO: Error message
+
+				profileUpdates.pinnedGalleryPostId = galleryPost.id;
+			} else if (ps.pinnedGalleryPostId === null) {
+				profileUpdates.pinnedGalleryPostId = null;
+			}
+
 			if (ps.fields) {
 				profileUpdates.fields = ps.fields
 					.filter(x => typeof x.name === 'string' && x.name.trim() !== '' && typeof x.value === 'string' && x.value.trim() !== '')
diff --git a/packages/backend/src/server/web/ClientServerService.ts b/packages/backend/src/server/web/ClientServerService.ts
index f8b3843cac..84694f13b8 100644
--- a/packages/backend/src/server/web/ClientServerService.ts
+++ b/packages/backend/src/server/web/ClientServerService.ts
@@ -375,6 +375,10 @@ export class ClientServerService {
 			return reply.sendFile('/favicon.ico', staticAssets);
 		});
 
+		fastify.get('/favicon.png', async (request, reply) => {
+			return reply.sendFile('/favicon.png', staticAssets);
+		});
+
 		fastify.get('/apple-touch-icon.png', async (request, reply) => {
 			return reply.sendFile('/apple-touch-icon.png', staticAssets);
 		});
@@ -639,9 +643,9 @@ export class ClientServerService {
 				relations: ['user'],
 			});
 
-			if (note && !note.user!.requireSigninToViewContents) {
+			if (note && (note.anonymouslySendToUserId || note.anonymousChannelUsername || !note.user!.requireSigninToViewContents)) {
 				const _note = await this.noteEntityService.pack(note);
-				const profile = await this.userProfilesRepository.findOneByOrFail({ userId: note.userId });
+				const profile = await this.userProfilesRepository.findOneByOrFail({ userId: _note.userId });
 				reply.header('Cache-Control', 'public, max-age=15');
 				if (profile.preventAiLearning) {
 					reply.header('X-Robots-Tag', 'noimageai');
diff --git a/packages/backend/src/server/web/FeedService.ts b/packages/backend/src/server/web/FeedService.ts
index 9d810ddc84..936abbea32 100644
--- a/packages/backend/src/server/web/FeedService.ts
+++ b/packages/backend/src/server/web/FeedService.ts
@@ -53,6 +53,7 @@ export class FeedService {
 				userId: user.id,
 				renoteId: IsNull(),
 				visibility: In(['public', 'home']),
+				anonymouslySendToUserId: IsNull(),
 			},
 			order: { id: -1 },
 			take: 20,
diff --git a/packages/backend/src/server/web/boot.embed.js b/packages/backend/src/server/web/boot.embed.js
index 9de1275380..1dc4f8fd4f 100644
--- a/packages/backend/src/server/web/boot.embed.js
+++ b/packages/backend/src/server/web/boot.embed.js
@@ -152,7 +152,7 @@
 			box-sizing: border-box;
 			overflow: hidden;
 
-			border-radius: var(--radius, 12px);
+			border-radius: var(--MI-radius, 12px);
 			border: 1px solid rgba(231, 255, 251, 0.14);
 		}
 
@@ -164,7 +164,7 @@
 			width: 100%;
 			height: 100%;
 			background: #192320;
-			border-radius: var(--radius, 12px);
+			border-radius: var(--MI-radius, 12px);
 			z-index: -1;
 		}
 
diff --git a/packages/backend/src/server/web/boot.js b/packages/backend/src/server/web/boot.js
index b55d327f86..ebc838221c 100644
--- a/packages/backend/src/server/web/boot.js
+++ b/packages/backend/src/server/web/boot.js
@@ -133,11 +133,12 @@
 	}
 
 	const customCss = localStorage.getItem('customCss');
-	if (customCss && customCss.length > 0) {
-		const style = document.createElement('style');
+	const style = document.createElement('style');
+	style.id = 'customCss';
+	if (customCss) {
 		style.innerHTML = customCss;
-		document.head.appendChild(style);
 	}
+	document.head.appendChild(style);
 
 	async function addStyle(styleText) {
 		let css = document.createElement('style');
diff --git a/packages/backend/src/server/web/manifest.json b/packages/backend/src/server/web/manifest.json
index 41171d62a1..3c86865801 100644
--- a/packages/backend/src/server/web/manifest.json
+++ b/packages/backend/src/server/web/manifest.json
@@ -1,10 +1,10 @@
 {
-	"short_name": "Misskey",
-	"name": "Misskey",
+	"short_name": "LTN",
+	"name": "LIVE THE@TER NET",
 	"start_url": "/",
 	"display": "standalone",
 	"background_color": "#313a42",
-	"theme_color": "#86b300",
+	"theme_color": "#ffc20b",
 	"icons": [
 		{
 			"src": "/static-assets/icons/192.png",
diff --git a/packages/backend/src/server/web/style.embed.css b/packages/backend/src/server/web/style.embed.css
index 5e8786cc4e..96a99d99dc 100644
--- a/packages/backend/src/server/web/style.embed.css
+++ b/packages/backend/src/server/web/style.embed.css
@@ -32,7 +32,7 @@ html.embed {
 html.embed #splash {
 	box-sizing: border-box;
 	min-height: 300px;
-	border-radius: var(--radius, 12px);
+	border-radius: var(--MI-radius, 12px);
 	border: 1px solid var(--MI_THEME-divider, #e8e8e8);
 }
 
diff --git a/packages/backend/src/server/web/views/base.pug b/packages/backend/src/server/web/views/base.pug
index 3883b5e5ab..28e3aa7c44 100644
--- a/packages/backend/src/server/web/views/base.pug
+++ b/packages/backend/src/server/web/views/base.pug
@@ -30,7 +30,7 @@ html
 		meta(property='instance_url' content= instanceUrl)
 		meta(name='viewport' content='width=device-width, initial-scale=1')
 		meta(name='format-detection' content='telephone=no,date=no,address=no,email=no,url=no')
-		link(rel='icon' href= icon || '/favicon.ico')
+		link(rel='icon' href= icon || '/favicon.png')
 		link(rel='apple-touch-icon' href= appleTouchIcon || '/apple-touch-icon.png')
 		link(rel='manifest' href='/manifest.json')
 		link(rel='search' type='application/opensearchdescription+xml' title=(title || "Misskey") href=`${baseUrl}/opensearch.xml`)
diff --git a/packages/frontend/src/components/MkNoteMediaGrid.vue b/packages/frontend/src/components/MkNoteMediaGrid.vue
index e51ea5a2de..89760de43f 100644
--- a/packages/frontend/src/components/MkNoteMediaGrid.vue
+++ b/packages/frontend/src/components/MkNoteMediaGrid.vue
@@ -4,7 +4,6 @@ SPDX-License-Identifier: AGPL-3.0-only
 -->
 
 <template>
-<template v-for="file in note.files">
 	<div
 		v-if="(((
 				(defaultStore.state.nsfw === 'force' || file.isSensitive) &&
@@ -31,7 +30,7 @@ SPDX-License-Identifier: AGPL-3.0-only
 			</div>
 		</div>
 	</div>
-	<MkA v-else :class="[$style.filePreview, { [$style.square]: square }]" :to="notePage(note)">
+	<MkA v-else :class="[$style.filePreview, { [$style.square]: square }]" :to="to">
 		<MkDriveFileThumbnail
 			:file="file"
 			fit="cover"
@@ -41,11 +40,9 @@ SPDX-License-Identifier: AGPL-3.0-only
 		/>
 	</MkA>
 </template>
-</template>
 
 <script lang="ts" setup>
 import { ref } from 'vue';
-import { notePage } from '@/filters/note.js';
 import { i18n } from '@/i18n.js';
 import * as Misskey from 'misskey-js';
 import { defaultStore } from '@/store.js';
@@ -54,7 +51,8 @@ import bytes from '@/filters/bytes.js';
 import MkDriveFileThumbnail from '@/components/MkDriveFileThumbnail.vue';
 
 defineProps<{
-	note: Misskey.entities.Note;
+	file: Misskey.entities.DriveFile;
+	to: string;
 	square?: boolean;
 }>();
 
diff --git a/packages/frontend/src/pages/admin/index.vue b/packages/frontend/src/pages/admin/index.vue
index 672ee8da18..96e779b4d8 100644
--- a/packages/frontend/src/pages/admin/index.vue
+++ b/packages/frontend/src/pages/admin/index.vue
@@ -9,7 +9,7 @@ SPDX-License-Identifier: AGPL-3.0-only
 		<MkSpacer :contentMax="700" :marginMin="16">
 			<div class="lxpfedzu _gaps">
 				<div class="banner">
-					<img :src="instance.iconUrl || '/favicon.ico'" alt="" class="icon"/>
+					<img :src="instance.iconUrl || '/favicon.png'" alt="" class="icon"/>
 				</div>
 
 				<div class="_gaps_s">
diff --git a/packages/frontend/src/pages/admin/security.vue b/packages/frontend/src/pages/admin/security.vue
index 975a4a1265..1f40a1af45 100644
--- a/packages/frontend/src/pages/admin/security.vue
+++ b/packages/frontend/src/pages/admin/security.vue
@@ -101,6 +101,19 @@ SPDX-License-Identifier: AGPL-3.0-only
 				</div>
 			</MkFolder>
 
+			<MkFolder>
+				<template #label>Banned Media Domains</template>
+				<template v-if="bannedMediaDomainsForm.modified.value" #footer>
+					<MkFormFooter :form="bannedEmailDomainsForm"/>
+				</template>
+
+				<div class="_gaps_m">
+					<MkTextarea v-model="bannedMediaDomainsForm.state.bannedMediaDomains">
+						<template #label>Banned Media Domains List</template>
+					</MkTextarea>
+				</div>
+			</MkFolder>
+
 			<MkFolder>
 				<template #label>Log IP address</template>
 				<template v-if="ipLoggingForm.savedState.enableIpLogging" #suffix>Enabled</template>
@@ -202,6 +215,15 @@ const bannedEmailDomainsForm = useForm({
 	fetchInstance(true);
 });
 
+const bannedMediaDomainsForm = useForm({
+	bannedMediaDomains: meta.bannedMediaDomains?.join('\n') || '',
+}, async (state) => {
+	await os.apiWithDialog('admin/update-meta', {
+		bannedMediaDomains: state.bannedMediaDomains.split('\n'),
+	});
+	fetchInstance(true);
+});
+
 const headerActions = computed(() => []);
 
 const headerTabs = computed(() => []);
diff --git a/packages/frontend/src/pages/gallery/post.vue b/packages/frontend/src/pages/gallery/post.vue
index feb4c60611..51a11741ae 100644
--- a/packages/frontend/src/pages/gallery/post.vue
+++ b/packages/frontend/src/pages/gallery/post.vue
@@ -35,7 +35,7 @@ SPDX-License-Identifier: AGPL-3.0-only
 							</div>
 						</div>
 						<div class="user">
-							<MkAvatar :user="post.user" class="avatar" link preview/>
+							<MkAvatar :user="post.user" class="avatar" link preview small/>
 							<div class="name">
 								<MkUserName :user="post.user" style="display: block;"/>
 								<MkAcct :user="post.user"/>
@@ -204,7 +204,19 @@ function showMenu(ev: MouseEvent) {
 
 watch(() => props.postId, fetchPost, { immediate: true });
 
-const headerActions = computed(() => []);
+const headerActions = computed(() => [{
+	icon: 'ti ti-pin',
+	text: i18n.ts.pin,
+	handler: () => {
+		os.apiWithDialog('i/update', {
+			pinnedGalleryPostId: props.postId,
+		});
+	},
+}, {
+	icon: 'ti ti-pencil',
+	text: i18n.ts.edit,
+	handler: edit,
+}]);
 
 const headerTabs = computed(() => []);
 
diff --git a/packages/frontend/src/pages/settings/profile.vue b/packages/frontend/src/pages/settings/profile.vue
index 561894d2b7..99132a0658 100644
--- a/packages/frontend/src/pages/settings/profile.vue
+++ b/packages/frontend/src/pages/settings/profile.vue
@@ -16,6 +16,7 @@ SPDX-License-Identifier: AGPL-3.0-only
 				<MkButton primary rounded link to="/settings/avatar-decoration">{{ i18n.ts.decorate }} <i class="ti ti-sparkles"></i></MkButton>
 			</div>
 		</div>
+		<MkButton v-if="hasPinnedGalleryPost" primary rounded :class="$style.galleryPostRemove" @click="removePinnedGalleryPost">{{ i18n.ts.unpin }} ({{ i18n.ts.gallery }})</MkButton>
 	</div>
 
 	<MkInput v-model="profile.name" :max="30" manualSave :mfmAutocomplete="['emoji']">
@@ -141,6 +142,7 @@ const $i = signinRequired();
 const Sortable = defineAsyncComponent(() => import('vuedraggable').then(x => x.default));
 
 const reactionAcceptance = computed(defaultStore.makeGetterSetter('reactionAcceptance'));
+const hasPinnedGalleryPost = ref(false);
 
 function assertVaildLang(lang: string | null): lang is keyof typeof langmap {
 	return lang != null && lang in langmap;
@@ -166,6 +168,13 @@ watch(() => profile, () => {
 const fields = ref($i.fields.map(field => ({ id: Math.random().toString(), name: field.name, value: field.value })) ?? []);
 const fieldEditMode = ref(false);
 
+function removePinnedGalleryPost() {
+	os.apiWithDialog('i/update', {
+		pinnedGalleryPostId: null,
+	});
+	hasPinnedGalleryPost.value = false;
+}
+
 function addField() {
 	fields.value.push({
 		id: Math.random().toString(),
@@ -314,6 +323,12 @@ definePageMetadata(() => ({
 	right: 16px;
 }
 
+.galleryPostRemove {
+	position: absolute;
+	bottom: 16px;
+	right: 16px;
+}
+
 .metadataRoot {
 	container-type: inline-size;
 }
diff --git a/packages/frontend/src/pages/user/files.vue b/packages/frontend/src/pages/user/files.vue
index b6c7c1c777..96a34b46d1 100644
--- a/packages/frontend/src/pages/user/files.vue
+++ b/packages/frontend/src/pages/user/files.vue
@@ -8,7 +8,9 @@ SPDX-License-Identifier: AGPL-3.0-only
 		<div :class="$style.root">
 			<MkPagination v-slot="{items}" :pagination="pagination">
 				<div :class="$style.stream">
-					<MkNoteMediaGrid v-for="note in items" :note="note" square/>
+					<template v-for="note in items" :key="note.id">
+						<MkNoteMediaGrid v-for="file in note.files ?? []" :key="file.id" :file="file" :to="notePage(note)" square/>
+					</template>
 				</div>
 			</MkPagination>
 		</div>
@@ -21,6 +23,7 @@ import * as Misskey from 'misskey-js';
 
 import MkNoteMediaGrid from '@/components/MkNoteMediaGrid.vue';
 import MkPagination from '@/components/MkPagination.vue';
+import { notePage } from '@/filters/note.js';
 
 const props = defineProps<{
 	user: Misskey.entities.UserDetailed;
diff --git a/packages/frontend/src/pages/user/index.files.vue b/packages/frontend/src/pages/user/index.files.vue
index b5e5f29ade..2441f81655 100644
--- a/packages/frontend/src/pages/user/index.files.vue
+++ b/packages/frontend/src/pages/user/index.files.vue
@@ -9,10 +9,10 @@ SPDX-License-Identifier: AGPL-3.0-only
 	<template #header>{{ i18n.ts.files }}</template>
 	<div :class="$style.root">
 		<MkLoading v-if="fetching"/>
-		<div v-if="!fetching && notes.length > 0" :class="$style.stream">
-			<MkNoteMediaGrid v-for="note in notes" :note="note"/>
+		<div v-if="!fetching && files.length > 0" :class="$style.stream">
+			<MkNoteMediaGrid v-for="entry in files" :key="entry.file.id" :file="entry.file" :to="entry.to"/>
 		</div>
-		<p v-if="!fetching && notes.length == 0" :class="$style.empty">{{ i18n.ts.nothing }}</p>
+		<p v-if="!fetching && files.length == 0" :class="$style.empty">{{ i18n.ts.nothing }}</p>
 	</div>
 </MkContainer>
 </template>
@@ -21,6 +21,7 @@ SPDX-License-Identifier: AGPL-3.0-only
 import { onMounted, ref } from 'vue';
 import * as Misskey from 'misskey-js';
 import { misskeyApi } from '@/scripts/misskey-api.js';
+import { notePage } from '@/filters/note.js';
 import MkContainer from '@/components/MkContainer.vue';
 import { i18n } from '@/i18n.js';
 import MkNoteMediaGrid from '@/components/MkNoteMediaGrid.vue';
@@ -32,9 +33,11 @@ const props = defineProps<{
 const emit = defineEmits<{
 	(ev: 'unfold'): void;
 }>();
-
 const fetching = ref(true);
-const notes = ref<Misskey.entities.Note[]>([]);
+const files = ref<{
+	to: string;
+	file: Misskey.entities.DriveFile;
+}[]>([]);
 
 function unfoldContainer(): boolean {
 	emit('unfold');
@@ -42,12 +45,29 @@ function unfoldContainer(): boolean {
 }
 
 onMounted(() => {
+	if (props.user.pinnedGalleryPost) {
+		for (const file of props.user.pinnedGalleryPost.files) {
+			files.value.push({
+				to: `/gallery/${props.user.pinnedGalleryPost.id}`,
+				file,
+			});
+		}
+		fetching.value = false;
+		return;
+	}
 	misskeyApi('users/notes', {
 		userId: props.user.id,
 		withFiles: true,
-		limit: 10,
-	}).then(_notes => {
-		notes.value = _notes;
+		limit: 15,
+	}).then(notes => {
+		for (const note of notes) {
+			for (const file of note.files ?? []) {
+				files.value.push({
+					to: notePage(note),
+					file,
+				});
+			}
+		}
 		fetching.value = false;
 	});
 });
