diff --git a/packages/backend/migration/1715618371464-anonymousChannel.js b/packages/backend/migration/1715618371464-anonymousChannel.js
new file mode 100644
index 0000000000..58cb316cf6
--- /dev/null
+++ b/packages/backend/migration/1715618371464-anonymousChannel.js
@@ -0,0 +1,34 @@
+/*
+ * SPDX-FileCopyrightText: syuilo and other misskey contributors
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+export class AnonymousChannel1715618371464 {
+    name = 'AnonymousChannel1715618371464'
+
+    async up(queryRunner) {
+        await queryRunner.query(`CREATE TYPE "public"."channel_anonymousstrategy_enum" AS ENUM('daily', 'weekly', 'monthly', 'yearly', 'manual')`);
+        await queryRunner.query(`ALTER TABLE "channel" ADD "anonymousStrategy" "public"."channel_anonymousstrategy_enum"`);
+        await queryRunner.query(`ALTER TABLE "channel" ADD "requirePublicWriteAccess" boolean NOT NULL DEFAULT false`);
+        await queryRunner.query(`ALTER TABLE "note" ADD "anonymousChannelUsername" character varying(128)`);
+        await queryRunner.query(`CREATE TABLE "channel_anonymous_salt" ("id" character varying(32) NOT NULL, "channelId" character varying(32) NOT NULL, "since" character varying(32) NOT NULL, "until" character varying(32), "salt" bigint NOT NULL, CONSTRAINT "PK_b918eb2c39bc401382cf4450af3" PRIMARY KEY ("id"))`);
+        await queryRunner.query(`CREATE INDEX "IDX_37988dcb4783a496681dcd4b4a" ON "channel_anonymous_salt" ("channelId") `);
+        await queryRunner.query(`CREATE INDEX "IDX_b86b095617434a08bd1826489" ON "channel_anonymous_salt" ("since") `);
+        await queryRunner.query(`CREATE INDEX "IDX_9b6d30f18ed64a82ac5aa5fc0" ON "channel_anonymous_salt" ("until") `);
+        await queryRunner.query(`CREATE UNIQUE INDEX "IDX_f7be165bf5ef42328a171a8a1" ON "channel_anonymous_salt" ("channelId", "since", "until") `);
+        await queryRunner.query(`ALTER TABLE "channel_anonymous_salt" ADD CONSTRAINT "FK_52001e5bea6d49bd8a19243d3b" FOREIGN KEY ("channelId") REFERENCES "channel"("id") ON DELETE CASCADE ON UPDATE NO ACTION`);
+    }
+
+    async down(queryRunner) {
+        await queryRunner.query(`ALTER TABLE "channel_anonymous_salt" DROP CONSTRAINT "FK_52001e5bea6d49bd8a19243d3b"`);
+        await queryRunner.query(`DROP INDEX "IDX_f7be165bf5ef42328a171a8a1"`);
+        await queryRunner.query(`DROP INDEX "IDX_9b6d30f18ed64a82ac5aa5fc0"`);
+        await queryRunner.query(`DROP INDEX "IDX_b86b095617434a08bd1826489"`);
+        await queryRunner.query(`DROP INDEX "IDX_37988dcb4783a496681dcd4b4a"`);
+        await queryRunner.query(`DROP TABLE "channel_anonymous_salt"`);
+        await queryRunner.query(`ALTER TABLE "note" DROP COLUMN "anonymousChannelUsername"`);
+        await queryRunner.query(`ALTER TABLE "channel" DROP COLUMN "requirePublicWriteAccess"`);
+        await queryRunner.query(`ALTER TABLE "channel" DROP COLUMN "anonymousStrategy"`);
+        await queryRunner.query(`DROP TYPE "public"."channel_anonymousstrategy_enum"`);
+    }
+}
diff --git a/packages/backend/src/core/AntennaService.ts b/packages/backend/src/core/AntennaService.ts
index e827ffa68c..9987fa3c1c 100644
--- a/packages/backend/src/core/AntennaService.ts
+++ b/packages/backend/src/core/AntennaService.ts
@@ -120,6 +120,8 @@ export class AntennaService implements OnApplicationShutdown {
 
 		if (!antenna.withReplies && note.replyId != null) return false;
 
+		if (note.anonymouslySendToUserId) return false;
+		if (note.anonymousChannelUsername) return false;
 		if (antenna.src === 'home') {
 			// TODO
 		} else if (antenna.src === 'list') {
diff --git a/packages/backend/src/core/FanoutTimelineEndpointService.ts b/packages/backend/src/core/FanoutTimelineEndpointService.ts
index b05af99c5e..31e3a152a1 100644
--- a/packages/backend/src/core/FanoutTimelineEndpointService.ts
+++ b/packages/backend/src/core/FanoutTimelineEndpointService.ts
@@ -68,7 +68,7 @@ export class FanoutTimelineEndpointService {
 
 		let noteIds = redisResultIds.slice(0, ps.limit);
 		const oldestNoteId = ascending ? redisResultIds[0] : redisResultIds[redisResultIds.length - 1];
-		const shouldFallbackToDb = noteIds.length === 0 || ps.sinceId != null && ps.sinceId < oldestNoteId;
+		const shouldFallbackToDb = noteIds.length === 0 || ps.sinceId != null && ps.sinceId < oldestNoteId || redisResult.some(timeline => !timeline.length);
 
 		if (!shouldFallbackToDb) {
 			let filter = ps.noteFilter ?? (_note => true);
@@ -110,9 +110,9 @@ export class FanoutTimelineEndpointService {
 
 				const parentFilter = filter;
 				filter = (note) => {
-					if (isUserRelated(note, userIdsWhoBlockingMe, ps.ignoreAuthorFromBlock)) return false;
-					if (isUserRelated(note, userIdsWhoMeMuting, ps.ignoreAuthorFromMute)) return false;
-					if (!ps.ignoreAuthorFromMute && isRenote(note) && !isQuote(note) && userIdsWhoMeMutingRenotes.has(note.userId)) return false;
+					if (isUserRelated(note, userIdsWhoBlockingMe, !!note.anonymousChannelUsername || ps.ignoreAuthorFromBlock)) return false;
+					if (isUserRelated(note, userIdsWhoMeMuting, !!note.anonymousChannelUsername || ps.ignoreAuthorFromMute)) return false;
+					if (!note.anonymousChannelUsername && !ps.ignoreAuthorFromMute && isRenote(note) && !isQuote(note) && userIdsWhoMeMutingRenotes.has(note.userId)) return false;
 					if (isInstanceMuted(note, userMutedInstances)) return false;
 
 					return parentFilter(note);
diff --git a/packages/backend/src/core/entities/ChannelEntityService.ts b/packages/backend/src/core/entities/ChannelEntityService.ts
index 1ba7ca8e57..314ce93d3a 100644
--- a/packages/backend/src/core/entities/ChannelEntityService.ts
+++ b/packages/backend/src/core/entities/ChannelEntityService.ts
@@ -82,9 +82,11 @@ export class ChannelEntityService {
 			pinnedNoteIds: channel.pinnedNoteIds,
 			color: channel.color,
 			isArchived: channel.isArchived,
+			anonymousStrategy: channel.anonymousStrategy,
 			usersCount: channel.usersCount,
 			notesCount: channel.notesCount,
 			isSensitive: channel.isSensitive,
+			requirePublicWriteAccess: channel.requirePublicWriteAccess,
 			allowRenoteToExternal: channel.allowRenoteToExternal,
 
 			...(me ? {
diff --git a/packages/backend/src/models/Channel.ts b/packages/backend/src/models/Channel.ts
index f5e9b17e3e..1fb04115b9 100644
--- a/packages/backend/src/models/Channel.ts
+++ b/packages/backend/src/models/Channel.ts
@@ -75,6 +75,17 @@ export class MiChannel {
 	})
 	public isArchived: boolean;
 
+	@Column('enum', {
+		enum: ['daily', 'weekly', 'monthly', 'yearly', 'manual'],
+		nullable: true,
+	})
+	public anonymousStrategy: 'daily' | 'weekly' | 'monthly' | 'yearly' | 'manual' | null;
+
+	@Column('boolean', {
+		default: false,
+	})
+	public requirePublicWriteAccess: boolean;
+
 	@Index()
 	@Column('integer', {
 		default: 0,
diff --git a/packages/backend/src/models/ChannelAnonymousSalt.ts b/packages/backend/src/models/ChannelAnonymousSalt.ts
new file mode 100644
index 0000000000..48b1175d53
--- /dev/null
+++ b/packages/backend/src/models/ChannelAnonymousSalt.ts
@@ -0,0 +1,43 @@
+/*
+ * SPDX-FileCopyrightText: syuilo and misskey-project
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import { PrimaryColumn, Entity, Index, JoinColumn, Column, ManyToOne } from 'typeorm';
+import { id } from './util/id.js';
+import { MiChannel } from './Channel.js';
+
+@Entity('channel_anonymous_salt')
+@Index(['channelId', 'since', 'until'], { unique: true })
+export class MiChannelAnonymousSalt {
+	@PrimaryColumn(id())
+	public id: string;
+
+	@Index()
+	@Column({
+		...id(),
+	})
+	public channelId: MiChannel['id'];
+
+	@ManyToOne(type => MiChannel, {
+		onDelete: 'CASCADE',
+	})
+	@JoinColumn()
+	public channel: MiChannel | null;
+
+	@Index()
+	@Column('varchar', {
+		...id(),
+	})
+	public since: string;
+
+	@Index()
+	@Column('varchar', {
+		...id(),
+		nullable: true,
+	})
+	public until: string | null;
+
+	@Column('bigint')
+	public salt: string;
+}
diff --git a/packages/backend/src/models/json-schema/channel.ts b/packages/backend/src/models/json-schema/channel.ts
index d233f7858d..dbb600ccc2 100644
--- a/packages/backend/src/models/json-schema/channel.ts
+++ b/packages/backend/src/models/json-schema/channel.ts
@@ -56,6 +56,11 @@ export const packedChannelSchema = {
 			type: 'boolean',
 			optional: false, nullable: false,
 		},
+		anonymousStrategy: {
+			type: 'string',
+			enum: ['daily', 'weekly', 'monthly', 'yearly', 'manual'],
+			nullable: true, optional: false,
+		},
 		usersCount: {
 			type: 'number',
 			nullable: false, optional: false,
diff --git a/packages/backend/src/server/api/endpoints/channels/create.ts b/packages/backend/src/server/api/endpoints/channels/create.ts
index e3a6d2d670..5468f27357 100644
--- a/packages/backend/src/server/api/endpoints/channels/create.ts
+++ b/packages/backend/src/server/api/endpoints/channels/create.ts
@@ -50,6 +50,8 @@ export const paramDef = {
 		bannerId: { type: 'string', format: 'misskey:id', nullable: true },
 		color: { type: 'string', minLength: 1, maxLength: 16 },
 		isSensitive: { type: 'boolean', nullable: true },
+		anonymousStrategy: { type: 'string', enum: ['daily', 'weekly', 'monthly', 'yearly', 'manual'], nullable: true },
+		requirePublicWriteAccess: { type: 'boolean', default: true },
 		allowRenoteToExternal: { type: 'boolean', nullable: true },
 	},
 	required: ['name'],
@@ -87,6 +89,8 @@ export default class extends Endpoint<typeof meta, typeof paramDef> { // eslint-
 				description: ps.description ?? null,
 				bannerId: banner ? banner.id : null,
 				isSensitive: ps.isSensitive ?? false,
+				anonymousStrategy: ps.anonymousStrategy ?? null,
+				requirePublicWriteAccess: ps.requirePublicWriteAccess,
 				...(ps.color !== undefined ? { color: ps.color } : {}),
 				allowRenoteToExternal: ps.allowRenoteToExternal ?? true,
 			} as MiChannel);
diff --git a/packages/backend/src/server/api/endpoints/channels/update.ts b/packages/backend/src/server/api/endpoints/channels/update.ts
index dba2938b39..5abc53c773 100644
--- a/packages/backend/src/server/api/endpoints/channels/update.ts
+++ b/packages/backend/src/server/api/endpoints/channels/update.ts
@@ -3,11 +3,13 @@
  * SPDX-License-Identifier: AGPL-3.0-only
  */
 
+import crypto from 'node:crypto';
 import { Inject, Injectable } from '@nestjs/common';
 import { Endpoint } from '@/server/api/endpoint-base.js';
-import type { DriveFilesRepository, ChannelsRepository } from '@/models/_.js';
+import type { DriveFilesRepository, ChannelsRepository, ChannelAnonymousSaltsRepository } from '@/models/_.js';
 import { ChannelEntityService } from '@/core/entities/ChannelEntityService.js';
 import { DI } from '@/di-symbols.js';
+import { IdService } from '@/core/IdService.js';
 import { RoleService } from '@/core/RoleService.js';
 import { ApiError } from '../../error.js';
 
@@ -53,6 +55,8 @@ export const paramDef = {
 		description: { type: 'string', nullable: true, minLength: 1, maxLength: 2048 },
 		bannerId: { type: 'string', format: 'misskey:id', nullable: true },
 		isArchived: { type: 'boolean', nullable: true },
+		requirePublicWriteAccess: { type: 'boolean', nullable: true },
+		regenerateAnonymousSalt: { type: 'boolean', default: false },
 		pinnedNoteIds: {
 			type: 'array',
 			items: {
@@ -72,11 +76,15 @@ export default class extends Endpoint<typeof meta, typeof paramDef> { // eslint-
 		@Inject(DI.channelsRepository)
 		private channelsRepository: ChannelsRepository,
 
+		@Inject(DI.channelAnonymousSaltsRepository)
+		private channelAnonymousSaltsRepository: ChannelAnonymousSaltsRepository,
+
 		@Inject(DI.driveFilesRepository)
 		private driveFilesRepository: DriveFilesRepository,
 
 		private channelEntityService: ChannelEntityService,
 
+		private idService: IdService,
 		private roleService: RoleService,
 	) {
 		super(meta, paramDef, async (ps, me) => {
@@ -108,12 +116,22 @@ export default class extends Endpoint<typeof meta, typeof paramDef> { // eslint-
 				banner = null;
 			}
 
+			if (ps.regenerateAnonymousSalt && channel.anonymousStrategy === 'manual') {
+				await this.channelAnonymousSaltsRepository.insert({
+					channelId: channel.id,
+					since: this.idService.gen(undefined, true),
+					until: null,
+					salt: crypto.getRandomValues(new BigUint64Array(1))[0].toString(),
+				});
+			}
+
 			await this.channelsRepository.update(channel.id, {
 				...(ps.name !== undefined ? { name: ps.name } : {}),
 				...(ps.description !== undefined ? { description: ps.description } : {}),
 				...(ps.pinnedNoteIds !== undefined ? { pinnedNoteIds: ps.pinnedNoteIds } : {}),
 				...(ps.color !== undefined ? { color: ps.color } : {}),
 				...(typeof ps.isArchived === 'boolean' ? { isArchived: ps.isArchived } : {}),
+				...(typeof ps.requirePublicWriteAccess === 'boolean' ? { requirePublicWriteAccess: ps.requirePublicWriteAccess } : {}),
 				...(banner ? { bannerId: banner.id } : {}),
 				...(typeof ps.isSensitive === 'boolean' ? { isSensitive: ps.isSensitive } : {}),
 				...(typeof ps.allowRenoteToExternal === 'boolean' ? { allowRenoteToExternal: ps.allowRenoteToExternal } : {}),
diff --git a/packages/backend/src/server/api/stream/channel.ts b/packages/backend/src/server/api/stream/channel.ts
index 686aea423c..ea8f2c9714 100644
--- a/packages/backend/src/server/api/stream/channel.ts
+++ b/packages/backend/src/server/api/stream/channel.ts
@@ -67,12 +67,12 @@ export default abstract class Channel {
 		if (isInstanceMuted(note, new Set<string>(this.userProfile?.mutedInstances ?? []))) return true;
 
 		// 流れてきたNoteがミュートしているユーザーが関わる
-		if (isUserRelated(note, this.userIdsWhoMeMuting)) return true;
+		if (isUserRelated(note, this.userIdsWhoMeMuting, !!note.anonymousChannelUsername)) return true;
 		// 流れてきたNoteがブロックされているユーザーが関わる
-		if (isUserRelated(note, this.userIdsWhoBlockingMe)) return true;
+		if (isUserRelated(note, this.userIdsWhoBlockingMe, !!note.anonymousChannelUsername)) return true;
 
 		// 流れてきたNoteがリノートをミュートしてるユーザが行ったもの
-		if (isRenotePacked(note) && !isQuotePacked(note) && this.userIdsWhoMeMutingRenotes.has(note.user.id)) return true;
+		if (!note.anonymousChannelUsername && isRenotePacked(note) && !isQuotePacked(note) && this.userIdsWhoMeMutingRenotes.has(note.user.id)) return true;
 
 		return false;
 	}
diff --git a/packages/frontend/src/cache.ts b/packages/frontend/src/cache.ts
index bfe8fbe0e4..b3e58fedfb 100644
--- a/packages/frontend/src/cache.ts
+++ b/packages/frontend/src/cache.ts
@@ -12,3 +12,4 @@ export const rolesCache = new Cache(1000 * 60 * 30, () => misskeyApi('admin/role
 export const userListsCache = new Cache<Misskey.entities.UserList[]>(1000 * 60 * 30, () => misskeyApi('users/lists/list'));
 export const antennasCache = new Cache<Misskey.entities.Antenna[]>(1000 * 60 * 30, () => misskeyApi('antennas/list'));
 export const favoritedChannelsCache = new Cache<Misskey.entities.Channel[]>(1000 * 60 * 30, () => misskeyApi('channels/my-favorites', { limit: 100 }));
+export const followedChannelsCache = new Cache<Misskey.entities.Channel[]>(1000 * 60 * 30, () => misskeyApi('channels/followed', { limit: 100 }));
diff --git a/packages/frontend/src/components/MkChannelFollowButton.vue b/packages/frontend/src/components/MkChannelFollowButton.vue
index d4e4f6179a..c139b8ad18 100644
--- a/packages/frontend/src/components/MkChannelFollowButton.vue
+++ b/packages/frontend/src/components/MkChannelFollowButton.vue
@@ -29,6 +29,7 @@ import { ref } from 'vue';
 import * as Misskey from 'misskey-js';
 import { misskeyApi } from '@/scripts/misskey-api.js';
 import { i18n } from '@/i18n.js';
+import { followedChannelsCache } from '@/cache.js';
 
 const props = withDefaults(defineProps<{
 	channel: Misskey.entities.Channel;
@@ -55,6 +56,7 @@ async function onClick() {
 			});
 			isFollowing.value = true;
 		}
+		followedChannelsCache.delete();
 	} catch (err) {
 		console.error(err);
 	} finally {
diff --git a/packages/frontend/src/components/MkReactionsViewer.reaction.vue b/packages/frontend/src/components/MkReactionsViewer.reaction.vue
index 41e475eade..439970ad4d 100644
--- a/packages/frontend/src/components/MkReactionsViewer.reaction.vue
+++ b/packages/frontend/src/components/MkReactionsViewer.reaction.vue
@@ -8,22 +8,28 @@ SPDX-License-Identifier: AGPL-3.0-only
 	ref="buttonEl"
 	v-ripple="canToggle"
 	class="_button"
-	:class="[$style.root, { [$style.reacted]: note.myReaction == reaction, [$style.canToggle]: canToggle, [$style.small]: defaultStore.state.reactionsDisplaySize === 'small', [$style.large]: defaultStore.state.reactionsDisplaySize === 'large' }]"
+	:class="[$style.root, { [$style.reacted]: note.myReaction == reaction, [$style.canToggle]: canToggle, [$style.small]: defaultStore.state.reactionsDisplaySize === 'small', [$style.large]: defaultStore.state.reactionsDisplaySize === 'large', [$style.x2]: conspicuousScale === 2 }]"
 	@click="toggleReaction()"
 	@contextmenu.prevent.stop="menu"
 >
 	<MkReactionIcon :class="defaultStore.state.limitWidthOfReaction ? $style.limitWidth : ''" :reaction="reaction" :emojiUrl="note.reactionEmojis[reaction.substring(1, reaction.length - 1)]"/>
 	<span :class="$style.count">{{ count }}</span>
+	<TransitionGroup v-if="props.users" :name="defaultStore.state.animation ? '_transition_zoom' : ''" :class="$style.users" tag="div">
+		<div v-for="(user, index) in props.users" :key="user.id" :class="[$style.user, { [$style.square]: defaultStore.state.squareAvatars }]" :style="{ zIndex: props.users.length - index }">
+			<MkAvatar :class="$style.avatar" :user="user" small forceHideDecoration/>
+		</div>
+	</TransitionGroup>
 </button>
 </template>
 
 <script lang="ts" setup>
-import { computed, inject, onMounted, shallowRef, watch } from 'vue';
+import { computed, inject, onBeforeUnmount, onMounted, shallowRef, watch } from 'vue';
 import * as Misskey from 'misskey-js';
 import { getUnicodeEmoji } from '@@/js/emojilist.js';
 import MkCustomEmojiDetailedDialog from './MkCustomEmojiDetailedDialog.vue';
 import XDetails from '@/components/MkReactionsViewer.details.vue';
 import MkReactionIcon from '@/components/MkReactionIcon.vue';
+import { globalEvents } from '@/events.js';
 import * as os from '@/os.js';
 import { misskeyApi, misskeyApiGet } from '@/scripts/misskey-api.js';
 import { useTooltip } from '@/scripts/use-tooltip.js';
@@ -41,6 +47,7 @@ const props = defineProps<{
 	count: number;
 	isInitial: boolean;
 	note: Misskey.entities.Note;
+	users?: Misskey.entities.UserLite[];
 }>();
 
 const mock = inject<boolean>('mock', false);
@@ -53,16 +60,30 @@ const buttonEl = shallowRef<HTMLElement>();
 
 const emojiName = computed(() => props.reaction.replace(/:/g, '').replace(/@\./, ''));
 const emoji = computed(() => customEmojisMap.get(emojiName.value) ?? getUnicodeEmoji(props.reaction));
+const conspicuousScale = computed(() => defaultStore.state.adaptiveReactionsDisplaySize && emoji.value.conspicuousScale || 1);
 
 const canToggle = computed(() => {
 	return !props.reaction.match(/@\w/) && $i && emoji.value && checkReactionPermissions($i, props.note, emoji.value);
 });
 const canGetInfo = computed(() => !props.reaction.match(/@\w/) && props.reaction.includes(':'));
 
+const abortController = new AbortController();
+
+onBeforeUnmount(() => abortController.abort());
+
 async function toggleReaction() {
 	if (!canToggle.value) return;
-
 	const oldReaction = props.note.myReaction;
+	if (props.note.anonymousChannelUsername && oldReaction !== props.reaction) {
+		const { canceled } = await os.confirm({
+			type: 'warning',
+			title: i18n.ts.reactionsAreNotAnonymized,
+			text: i18n.ts.areYouSure,
+		});
+		if (canceled) {
+			return;
+		}
+	}
 	if (oldReaction) {
 		const confirm = await os.confirm({
 			type: 'warning',
@@ -135,6 +156,8 @@ async function menu(ev) {
 }
 
 function anime() {
+	globalEvents.emit('appendReaction', props.reaction);
+
 	if (document.hidden || !defaultStore.state.animation || buttonEl.value == null) return;
 
 	const rect = buttonEl.value.getBoundingClientRect();
@@ -155,19 +178,17 @@ onMounted(() => {
 
 if (!mock) {
 	useTooltip(buttonEl, async (showing) => {
-		const reactions = await misskeyApiGet('notes/reactions', {
-			noteId: props.note.id,
-			type: props.reaction,
-			limit: 10,
-			_cacheKey_: props.count,
-		});
-
-		const users = reactions.map(x => x.user);
-
 		const { dispose } = os.popup(XDetails, {
 			showing,
 			reaction: props.reaction,
-			users,
+			users: props.users
+				? props.users.slice(0, 10)
+				: (await misskeyApiGet('notes/reactions', {
+					noteId: props.note.id,
+					type: props.reaction,
+					limit: 10,
+					_cacheKey_: props.count,
+				})).map(x => x.user),
 			count: props.count,
 			targetElement: buttonEl.value,
 		}, {
@@ -186,13 +207,17 @@ if (!mock) {
 	font-size: 1.5em;
 	border-radius: 6px;
 	align-items: center;
-	justify-content: center;
+	white-space: nowrap;
+	overflow: clip;
+	contain: paint;
 
 	&.canToggle {
 		background: var(--MI_THEME-buttonBg);
+		--x-bg: var(--MI_THEME-buttonBg);
 
 		&:hover {
 			background: rgba(0, 0, 0, 0.1);
+			--x-bg: rgba(0, 0, 0, 0.1);
 		}
 	}
 
@@ -200,6 +225,21 @@ if (!mock) {
 		cursor: default;
 	}
 
+	&.x2 {
+		height: 88px;
+		font-size: 4em;
+		float: left;
+
+		> .count {
+			font-size: 0.2625em;
+		}
+
+		> .users {
+			grid-template-rows: repeat(4, 16px);
+			margin: 0 8px 8px;
+		}
+	}
+
 	&.small {
 		height: 32px;
 		font-size: 1em;
@@ -209,6 +249,19 @@ if (!mock) {
 			font-size: 0.9em;
 			line-height: 32px;
 		}
+
+		&.x2 {
+			height: 68px;
+			font-size: 3em;
+
+			> .count {
+				font-size: 0.3em;
+			}
+
+			> .users {
+				grid-template-rows: repeat(3, 16px);
+			}
+		}
 	}
 
 	&.large {
@@ -220,10 +273,24 @@ if (!mock) {
 			font-size: 0.6em;
 			line-height: 52px;
 		}
+
+		&.x2 {
+			height: 108px;
+			font-size: 5em;
+
+			> .count {
+				font-size: 0.24em;
+			}
+
+			> .users {
+				grid-template-rows: repeat(5, 16px);
+			}
+		}
 	}
 
 	&.reacted, &.reacted:hover {
 		background: var(--MI_THEME-accentedBg);
+		--x-bg: var(--MI_THEME-accentedBg);
 		color: var(--MI_THEME-accent);
 		box-shadow: 0 0 0 1px var(--MI_THEME-accent) inset;
 
@@ -247,4 +314,44 @@ if (!mock) {
 	line-height: 42px;
 	margin: 0 0 0 4px;
 }
+
+.users {
+	display: inline-grid;
+	grid-auto-flow: column;
+	grid-auto-columns: 16px;
+	vertical-align: middle;
+	margin: 0 8px;
+}
+
+.user {
+	width: 24px;
+	height: 24px;
+	outline: 3px var(--MI_THEME-panel) solid;
+	border-radius: 100%;
+	background: var(--MI_THEME-panel) linear-gradient(0deg, var(--x-bg), var(--x-bg));
+
+	&::before {
+		content: "";
+		position: absolute;
+		display: block;
+		width: 24px;
+		height: 24px;
+		outline: 3px var(--x-bg) solid;
+		border-radius: 100%;
+	}
+
+	&.square {
+		border-radius: 20%;
+
+		&::before {
+			border-radius: 20%;
+		}
+	}
+}
+
+.avatar {
+	display: block;
+	width: 24px;
+	height: 24px;
+}
 </style>
diff --git a/packages/frontend/src/components/MkTimeline.vue b/packages/frontend/src/components/MkTimeline.vue
index 7bae240ddd..4654c1a868 100644
--- a/packages/frontend/src/components/MkTimeline.vue
+++ b/packages/frontend/src/components/MkTimeline.vue
@@ -22,6 +22,7 @@ import * as Misskey from 'misskey-js';
 import type { BasicTimelineType } from '@/timelines.js';
 import MkNotes from '@/components/MkNotes.vue';
 import MkPullToRefresh from '@/components/MkPullToRefresh.vue';
+import { globalEvents } from '@/events.js';
 import { useStream } from '@/stream.js';
 import * as sound from '@/scripts/sound.js';
 import { $i } from '@/account.js';
@@ -54,7 +55,7 @@ const emit = defineEmits<{
 
 provide('inTimeline', true);
 provide('tl_withSensitive', computed(() => props.withSensitive));
-provide('inChannel', computed(() => props.src === 'channel'));
+provide('inChannel', computed(() => props.src === 'channel' && props.channel || null));
 
 type TimelineQueryType = {
 	antennaId?: string,
@@ -73,6 +74,8 @@ const tlComponent = shallowRef<InstanceType<typeof MkNotes>>();
 let tlNotesCount = 0;
 
 function prepend(note) {
+	globalEvents.emit('prependTimeline', note);
+
 	if (tlComponent.value == null) return;
 
 	tlNotesCount++;
diff --git a/packages/frontend/src/local-storage.ts b/packages/frontend/src/local-storage.ts
index 9c8863f863..1726b71bb6 100644
--- a/packages/frontend/src/local-storage.ts
+++ b/packages/frontend/src/local-storage.ts
@@ -32,6 +32,7 @@ export type Keys = (
 	'message_drafts' |
 	'scratchpad' |
 	'debug' |
+	`channel:${string}` |
 	`miux:${string}` |
 	`ui:folder:${string}` |
 	`themes:${string}` |
diff --git a/packages/frontend/src/pages/channel-editor.vue b/packages/frontend/src/pages/channel-editor.vue
index 6d8274a55c..266b036bba 100644
--- a/packages/frontend/src/pages/channel-editor.vue
+++ b/packages/frontend/src/pages/channel-editor.vue
@@ -28,6 +28,15 @@ SPDX-License-Identifier: AGPL-3.0-only
 				<template #label>{{ i18n.ts._channel.allowRenoteToExternal }}</template>
 			</MkSwitch>
 
+			<MkSwitch v-model="requirePublicWriteAccess">
+				<template #label>{{ i18n.ts._channel.requirePublicWriteAccess }}</template>
+			</MkSwitch>
+
+			<MkSelect v-model="anonymousStrategy" :readonly="channelId != null">
+				<template #label>{{ i18n.ts._channel.anonymousStrategy }}</template>
+				<option v-for="item in ['none', 'daily', 'weekly', 'monthly', 'yearly', 'manual']" :key="item" :value="item">{{ i18n.ts[item] }}</option>
+			</MkSelect>
+
 			<div>
 				<MkButton v-if="bannerId == null" @click="setBannerImage"><i class="ti ti-plus"></i> {{ i18n.ts._channel.setBanner }}</MkButton>
 				<div v-else-if="bannerUrl">
@@ -62,6 +71,7 @@ SPDX-License-Identifier: AGPL-3.0-only
 			<div class="_buttons">
 				<MkButton primary @click="save()"><i class="ti ti-device-floppy"></i> {{ channelId ? i18n.ts.save : i18n.ts.create }}</MkButton>
 				<MkButton v-if="channelId" danger @click="archive()"><i class="ti ti-trash"></i> {{ i18n.ts.archive }}</MkButton>
+				<MkButton v-if="channel?.anonymousStrategy === 'manual'" danger @click="regenerateSalt()"><i class="ti ti-refresh-alert"></i> {{ i18n.ts.regenerate }}</MkButton>
 			</div>
 		</div>
 	</MkSpacer>
@@ -80,6 +90,7 @@ import { misskeyApi } from '@/scripts/misskey-api.js';
 import { definePageMetadata } from '@/scripts/page-metadata.js';
 import { i18n } from '@/i18n.js';
 import MkFolder from '@/components/MkFolder.vue';
+import MkSelect from '@/components/MkSelect.vue';
 import MkSwitch from '@/components/MkSwitch.vue';
 import MkTextarea from '@/components/MkTextarea.vue';
 import { useRouter } from '@/router/supplier.js';
@@ -100,6 +111,8 @@ const bannerId = ref<string | null>(null);
 const color = ref('#000');
 const isSensitive = ref(false);
 const allowRenoteToExternal = ref(true);
+const requirePublicWriteAccess = ref(true);
+const anonymousStrategy = ref<'none' | 'daily' | 'weekly' | 'monthly' | 'yearly' | 'manual'>('none');
 const pinnedNotes = ref<{ id: Misskey.entities.Note['id'] }[]>([]);
 
 watch(() => bannerId.value, async () => {
@@ -129,6 +142,8 @@ async function fetchChannel() {
 	}));
 	color.value = channel.value.color;
 	allowRenoteToExternal.value = channel.value.allowRenoteToExternal;
+	requirePublicWriteAccess.value = channel.value.requirePublicWriteAccess;
+	anonymousStrategy.value = channel.value.anonymousStrategy ?? 'none';
 }
 
 fetchChannel();
@@ -159,12 +174,16 @@ function save() {
 		color: color.value,
 		isSensitive: isSensitive.value,
 		allowRenoteToExternal: allowRenoteToExternal.value,
+		requirePublicWriteAccess: requirePublicWriteAccess.value,
 	};
 
 	if (props.channelId) {
 		params.channelId = props.channelId;
 		os.apiWithDialog('channels/update', params);
 	} else {
+		if (anonymousStrategy.value !== 'none') {
+			params.anonymousStrategy = anonymousStrategy.value;
+		}
 		os.apiWithDialog('channels/create', params).then(created => {
 			router.push(`/channels/${created.id}`);
 		});
@@ -188,6 +207,23 @@ async function archive() {
 	});
 }
 
+async function regenerateSalt() {
+	const { canceled } = await os.confirm({
+		type: 'warning',
+		title: i18n.ts.regenerate,
+		text: i18n.ts.areYouSure,
+	});
+
+	if (canceled) return;
+
+	misskeyApi('channels/update', {
+		channelId: props.channelId,
+		regenerateSalt: true,
+	}).then(() => {
+		os.success();
+	});
+}
+
 function setBannerImage(evt) {
 	selectFile(evt.currentTarget ?? evt.target, null).then(file => {
 		bannerId.value = file.id;
diff --git a/packages/frontend/src/pages/channel.vue b/packages/frontend/src/pages/channel.vue
index 4a91165d50..9eb5833d4c 100644
--- a/packages/frontend/src/pages/channel.vue
+++ b/packages/frontend/src/pages/channel.vue
@@ -32,6 +32,44 @@ SPDX-License-Identifier: AGPL-3.0-only
 						<MkNote v-for="note in channel.pinnedNotes" :key="note.id" class="_panel" :note="note"/>
 					</div>
 				</MkFoldableSection>
+				<MkFoldableSection :expanded="false">
+					<template #header><i class="ti ti-info-circle ti-fw" style="margin-right: 0.5em;"></i>{{ i18n.ts.info }}</template>
+					<FormSuspense v-slot="{ result: info }" :p="initInfo">
+						<FormSection>
+							<FormSplit>
+								<MkKeyValue>
+									<template #key>{{ i18n.ts.author }}</template>
+									<template #value>
+										<MkA v-if="info.user" :to="userPage(info.user)">
+											<MkUserCardMini :user="info.user"/>
+										</MkA>
+										<MkInfo v-else>{{ i18n.ts.none }}</MkInfo>
+									</template>
+								</MkKeyValue>
+								<MkKeyValue>
+									<template #key>{{ i18n.ts.createdAt }}</template>
+									<template #value><MkTime :time="channel.createdAt" colored/></template>
+								</MkKeyValue>
+								<MkKeyValue>
+									<template #key>{{ i18n.ts.updatedAt }}</template>
+									<template #value><MkTime :time="channel.lastNotedAt" colored/></template>
+								</MkKeyValue>
+								<MkKeyValue>
+									<template #key>{{ i18n.ts._channel.anonymousStrategy }}</template>
+									<template #value>{{ channel.anonymousStrategy ? i18n.ts[channel.anonymousStrategy] : i18n.ts.none }}</template>
+								</MkKeyValue>
+								<MkKeyValue>
+									<template #key>{{ i18n.ts._channel.requirePublicWriteAccess }}</template>
+									<template #value>{{ channel.requirePublicWriteAccess ? i18n.ts.yes : i18n.ts.no }}</template>
+								</MkKeyValue>
+								<MkKeyValue>
+									<template #key>{{ i18n.ts._channel.allowRenoteToExternal }}</template>
+									<template #value>{{ channel.allowRenoteToExternal ? i18n.ts.yes : i18n.ts.no }}</template>
+								</MkKeyValue>
+							</FormSplit>
+						</FormSection>
+					</FormSuspense>
+				</MkFoldableSection>
 			</div>
 			<div v-if="channel && tab === 'timeline'" key="timeline" class="_gaps">
 				<MkInfo v-if="channel.isArchived" warn>{{ i18n.ts.thisChannelArchived }}</MkInfo>
@@ -79,11 +117,17 @@ import MkPostForm from '@/components/MkPostForm.vue';
 import MkTimeline from '@/components/MkTimeline.vue';
 import XChannelFollowButton from '@/components/MkChannelFollowButton.vue';
 import * as os from '@/os.js';
+import { userPage } from '@/filters/user.js';
 import { misskeyApi } from '@/scripts/misskey-api.js';
 import { $i, iAmModerator } from '@/account.js';
 import { i18n } from '@/i18n.js';
 import { definePageMetadata } from '@/scripts/page-metadata.js';
 import { deviceKind } from '@/scripts/device-kind.js';
+import FormSection from '@/components/form/section.vue';
+import FormSplit from '@/components/form/split.vue';
+import FormSuspense from '@/components/form/suspense.vue';
+import MkKeyValue from '@/components/MkKeyValue.vue';
+import MkUserCardMini from '@/components/MkUserCardMini.vue';
 import MkNotes from '@/components/MkNotes.vue';
 import { url } from '@@/js/config.js';
 import { favoritedChannelsCache } from '@/cache.js';
@@ -121,11 +165,19 @@ const featuredPagination = computed(() => ({
 		channelId: props.channelId,
 	},
 }));
+const initInfo = () => channel.value?.userId
+	? misskeyApi('users/show', { userId: channel.value?.userId }).then(user => ({
+			user,
+		}))
+	: Promise.resolve({
+			user: null,
+		});
 
 watch(() => props.channelId, async () => {
 	channel.value = await misskeyApi('channels/show', {
 		channelId: props.channelId,
 	});
+	miLocalStorage.setItem(`channel:${props.channelId}`, JSON.stringify(channel.value));
 	favorited.value = channel.value.isFavorited ?? false;
 	if (favorited.value || channel.value.isFollowing) {
 		tab.value = 'timeline';
diff --git a/packages/frontend/src/pages/channels.vue b/packages/frontend/src/pages/channels.vue
index 6830c1ace4..772325c65a 100644
--- a/packages/frontend/src/pages/channels.vue
+++ b/packages/frontend/src/pages/channels.vue
@@ -60,7 +60,7 @@ SPDX-License-Identifier: AGPL-3.0-only
 </template>
 
 <script lang="ts" setup>
-import { computed, onMounted, ref } from 'vue';
+import { computed, onMounted, ref, watch } from 'vue';
 import MkChannelPreview from '@/components/MkChannelPreview.vue';
 import MkChannelList from '@/components/MkChannelList.vue';
 import MkPagination from '@/components/MkPagination.vue';
@@ -91,6 +91,12 @@ onMounted(() => {
 	searchType.value = props.type ?? 'nameAndDescription';
 });
 
+watch(tab, () => {
+	if (tab.value === 'search') {
+		search();
+	}
+});
+
 const featuredPagination = {
 	endpoint: 'channels/featured' as const,
 	limit: 10,
diff --git a/packages/frontend/src/types/menu.ts b/packages/frontend/src/types/menu.ts
index 7cadef136d..8ef6f9583d 100644
--- a/packages/frontend/src/types/menu.ts
+++ b/packages/frontend/src/types/menu.ts
@@ -13,11 +13,11 @@ export type MenuAction = (ev: MouseEvent) => void;
 export type MenuDivider = { type: 'divider' };
 export type MenuNull = undefined;
 export type MenuLabel = { type: 'label', text: string };
-export type MenuLink = { type: 'link', to: string, text: string, icon?: string, indicate?: boolean, avatar?: Misskey.entities.User };
-export type MenuA = { type: 'a', href: string, target?: string, download?: string, text: string, icon?: string, indicate?: boolean };
+export type MenuLink = { type: 'link', to: string, text: string, icon?: string, suffixIcon?: string, indicate?: boolean, avatar?: Misskey.entities.User };
+export type MenuA = { type: 'a', href: string, target?: string, download?: string, text: string, icon?: string, suffixIcon?: string, indicate?: boolean };
 export type MenuUser = { type: 'user', user: Misskey.entities.User, active?: boolean, indicate?: boolean, action: MenuAction };
 export type MenuSwitch = { type: 'switch', ref: Ref<boolean>, text: string, icon?: string, disabled?: boolean | Ref<boolean> };
-export type MenuButton = { type?: 'button', text: string, icon?: string, indicate?: boolean, danger?: boolean, active?: boolean | ComputedRef<boolean>, avatar?: Misskey.entities.User; action: MenuAction };
+export type MenuButton = { type?: 'button', text: string, icon?: string, suffixIcon?: string, indicate?: boolean, danger?: boolean, active?: boolean | ComputedRef<boolean>, avatar?: Misskey.entities.User; action: MenuAction };
 export type MenuRadio = { type: 'radio', text: string, icon?: string, ref: Ref<MenuRadioOptionsDef[keyof MenuRadioOptionsDef]>, options: MenuRadioOptionsDef, disabled?: boolean | Ref<boolean> };
 export type MenuRadioOption = { type: 'radioOption', text: string, action: MenuAction; active?: boolean | ComputedRef<boolean> };
 export type MenuParent = { type: 'parent', text: string, icon?: string, children: MenuItem[] | (() => Promise<MenuItem[]> | MenuItem[]) };
