diff --git a/.vscode/settings.json b/.vscode/settings.json
index 0ceec23acd..aa1b2b3dbf 100644
--- a/.vscode/settings.json
+++ b/.vscode/settings.json
@@ -9,7 +9,7 @@
 	"jest.jestCommandLine": "pnpm run jest",
 	"jest.runMode": "on-demand",
 	"editor.codeActionsOnSave": {
-		"source.fixAll": "explicit"
+		"source.fixAll": "never"
 	},
 	"editor.formatOnSave": false
 }
diff --git a/locales/en-US.yml b/locales/en-US.yml
index 6ff7e5fb7c..17c0fccf4a 100644
--- a/locales/en-US.yml
+++ b/locales/en-US.yml
@@ -20,6 +20,9 @@ cancel: "Cancel"
 noThankYou: "Not now"
 enterUsername: "Enter username"
 renotedBy: "Renoted by {user}"
+renotedByChannel: "Renoted by {user} to {channel}"
+publicLetter: "Public letter"
+anonymouslySendToUser: "Public letter adressed to {user}"
 noNotes: "No notes"
 noNotifications: "No notifications"
 instance: "Instance"
@@ -120,8 +123,8 @@ cantReRenote: "A renote can't be renoted."
 quote: "Quote"
 inChannelRenote: "Channel-only Renote"
 inChannelQuote: "Channel-only Quote"
-renoteToChannel: "Renote to channel"
-renoteToOtherChannel: "Renote to other channel"
+toChannelRenote: "Renote to channel"
+toChannelQuote: "Quote to channel"
 pinnedNote: "Pinned note"
 pinned: "Pin to profile"
 you: "You"
@@ -138,6 +141,7 @@ overwriteFromPinnedEmojisForReaction: "Override from reaction settings"
 overwriteFromPinnedEmojis: "Override from general settings"
 reactionSettingDescription2: "Drag to reorder, click to delete, press \"+\" to add."
 rememberNoteVisibility: "Remember note visibility settings"
+showNoteVisibilityDial: "Show note visibility dial"
 attachCancel: "Remove attachment"
 deleteFile: "Delete file"
 markAsSensitive: "Mark as sensitive"
@@ -376,6 +380,12 @@ today: "Today"
 dayX: "{day}"
 monthX: "{month}"
 yearX: "{year}"
+daily: "Daily"
+weekly: "Weekly"
+monthly: "Monthly"
+yearly: "Yearly"
+manual: "Manual"
+reactionsAreNotAnonymized: "Reactions are not anonymized"
 pages: "Pages"
 integration: "Integration"
 connectService: "Connect"
@@ -576,6 +586,8 @@ deleteAll: "Delete all"
 showFixedPostForm: "Display the posting form at the top of the timeline"
 showFixedPostFormInChannel: "Display the posting form at the top of the timeline (Channels)"
 withRepliesByDefaultForNewlyFollowed: "Include replies by newly followed users in the timeline by default"
+postToTheChannelYouHaveOpen: "Post to the channel you have open"
+autoOpenCWs: "CW that should be opened automatically"
 newNoteRecived: "There are new notes"
 sounds: "Sounds"
 sound: "Sounds"
@@ -881,12 +893,16 @@ high: "High"
 middle: "Medium"
 low: "Low"
 emailNotConfiguredWarning: "Email address not set."
+passkeyNotConfiguredWarning: "Passkey not set. If you see this message, the device you are using most likely can be a security key."
+twoFactorNotConfiguredWarning: "Two-factor authentication not set."
 ratio: "Ratio"
 previewNoteText: "Show preview"
 customCss: "Custom CSS"
 customCssWarn: "This setting should only be used if you know what it does. Entering improper values may cause the client to stop functioning normally."
 global: "Global"
 squareAvatars: "Display squared avatars"
+localNeighborNotes: "Local neighbor notes"
+globalNeighborNotes: "Global neighbor notes"
 sent: "Sent"
 received: "Received"
 searchResult: "Search results"
@@ -1098,8 +1114,11 @@ enableChartsForRemoteUser: "Generate remote user data charts"
 enableChartsForFederatedInstances: "Generate remote instance data charts"
 enableStatsForFederatedInstances: "Receive remote server stats"
 showClipButtonInNoteFooter: "Add \"Clip\" to note action menu"
+showTranslateButtonInNoteFooter: "Add \"Translate\" to note action menu"
 reactionsDisplaySize: "Reaction display size"
+adaptiveReactionsDisplaySize: "Adaptive size"
 limitWidthOfReaction: "Limit the maximum width of reactions and display them in reduced size."
+showReactedUserAvatars: "Show avatars of users who reacted"
 noteIdOrUrl: "Note ID or URL"
 video: "Video"
 videos: "Videos"
@@ -1783,6 +1802,7 @@ _role:
     canManageCustomEmojis: "Can manage custom emojis"
     canManageAvatarDecorations: "Manage avatar decorations"
     driveCapacity: "Drive capacity"
+    driveUploadBandwidth: "Drive upload bandwidth"
     alwaysMarkNsfw: "Always mark files as NSFW"
     canUpdateBioMedia: "Can edit an icon or a banner image"
     pinMax: "Maximum number of pinned notes"
@@ -1813,6 +1833,9 @@ _role:
     isSuspended: "Suspended user"
     isLocked: "Private accounts"
     isExplorable: "Effective user of \"make an account discoverable\""
+    isEmailVerified: "Email address verified"
+    isTwoFactorEnabled: "Two-factor authentication enabled"
+    isPasswordLessLoginEnabled: "Password-less login enabled"
     createdLessThan: "Less than X has passed since account creation"
     createdMoreThan: "More than X has passed since account creation"
     followersLessThanOrEq: "Has X or fewer followers"
@@ -1944,6 +1967,8 @@ _channel:
   nameAndDescription: "Name and description"
   nameOnly: "Name only"
   allowRenoteToExternal: "Allow renote and quote outside the channel"
+  requirePublicWriteAccess: "Require public write access"
+  anonymousStrategy: "Anonymous strategy"
 _menuDisplay:
   sideFull: "Side"
   sideIcon: "Side (Icons)"
@@ -2279,6 +2304,7 @@ _postForm:
   replyPlaceholder: "Reply to this note..."
   quotePlaceholder: "Quote this note..."
   channelPlaceholder: "Post to a channel..."
+  placeholder: "Post as {x}..."
   _placeholders:
     a: "What are you up to?"
     b: "What's happening around you?"
@@ -2700,6 +2726,7 @@ _reversi:
   disallowIrregularRules: "No irregular rules"
   showBoardLabels: "Display row and column numbering on the board"
   useAvatarAsStone: "Turn stones into user avatars"
+  use10secarisaAsStone: "Turn stone into 10sec-arisa"
 _offlineScreen:
   title: "Offline - cannot connect to the server"
   header: "Unable to connect to the server"
diff --git a/locales/index.d.ts b/locales/index.d.ts
index 0f71263c96..9a88089b41 100644
--- a/locales/index.d.ts
+++ b/locales/index.d.ts
@@ -98,6 +98,18 @@ export interface Locale extends ILocale {
      * {user}がリノート
      */
     "renotedBy": ParameterizedString<"user">;
+    /**
+     * {user}が{channel}にリノート
+     */
+    "renotedByChannel": ParameterizedString<"user" | "channel">;
+    /**
+     * パブリックレター
+     */
+    "publicLetter": string;
+    /**
+     * {user}宛のパブリックレター
+     */
+    "anonymouslySendToUser": ParameterizedString<"user">;
     /**
      * ノートはありません
      */
@@ -474,10 +486,6 @@ export interface Locale extends ILocale {
      * リノートしました。
      */
     "renoted": string;
-    /**
-     * {name} にリノートしました。
-     */
-    "renotedToX": ParameterizedString<"name">;
     /**
      * この投稿はリノートできません。
      */
@@ -499,13 +507,13 @@ export interface Locale extends ILocale {
      */
     "inChannelQuote": string;
     /**
-     * チャンネルにリノート
+     * チャンネルにRenote
      */
-    "renoteToChannel": string;
+    "toChannelRenote": string;
     /**
-     * 他のチャンネルにリノート
+     * チャンネルに引用
      */
-    "renoteToOtherChannel": string;
+    "toChannelQuote": string;
     /**
      * ピン留めされたノート
      */
@@ -570,6 +578,10 @@ export interface Locale extends ILocale {
      * 公開範囲を記憶する
      */
     "rememberNoteVisibility": string;
+    /**
+     * 公開範囲ダイアルを表示
+     */
+    "showNoteVisibilityDial": string;
     /**
      * 添付取り消し
      */
@@ -1522,6 +1534,30 @@ export interface Locale extends ILocale {
      * {year}年
      */
     "yearX": ParameterizedString<"year">;
+    /**
+     * 日次
+     */
+    "daily": string;
+    /**
+     * 週次
+     */
+    "weekly": string;
+    /**
+     * 月次
+     */
+    "monthly": string;
+    /**
+     * 年次
+     */
+    "yearly": string;
+    /**
+     * 手動
+     */
+    "manual": string;
+    /**
+     * リアクションは匿名化されません
+     */
+    "reactionsAreNotAnonymized": string;
     /**
      * ページ
      */
@@ -2322,6 +2358,14 @@ export interface Locale extends ILocale {
      * フォローする際、デフォルトで返信をTLに含むようにする
      */
     "withRepliesByDefaultForNewlyFollowed": string;
+    /**
+     * 開いているチャンネルに投稿する
+     */
+    "postToTheChannelYouHaveOpen": string;
+    /**
+     * 自動で開くCW
+     */
+    "autoOpenCws": string;
     /**
      * 新しいノートがあります
      */
@@ -3542,6 +3586,14 @@ export interface Locale extends ILocale {
      * メールアドレスの設定がされていません。
      */
     "emailNotConfiguredWarning": string;
+    /**
+     * セキュリティキー・パスキーの設定がされていません。このメッセージが表示されている場合、おそらくあなたが今使っているデバイスはパスキーとして使用できます。
+     */
+    "passkeyNotConfiguredWarning": string;
+    /**
+     * 二要素認証の設定がされていません。
+     */
+    "twoFactorNotConfiguredWarning": string;
     /**
      * 比率
      */
@@ -3566,6 +3618,14 @@ export interface Locale extends ILocale {
      * アイコンを四角形で表示
      */
     "squareAvatars": string;
+    /**
+     * ローカルの近隣ノート
+     */
+    "localNeighborNotes": string;
+    /**
+     * グローバルの近隣ノート
+     */
+    "globalNeighborNotes": string;
     /**
      * 送信
      */
@@ -4410,10 +4470,22 @@ export interface Locale extends ILocale {
      * ノートのアクションにクリップを追加
      */
     "showClipButtonInNoteFooter": string;
+    /**
+     * ノートのアクションに翻訳を追加
+     */
+    "showTranslateButtonInNoteFooter": string;
     /**
      * リアクションの表示サイズ
      */
     "reactionsDisplaySize": string;
+    /**
+     * 適応型サイズ
+     */
+    "adaptiveReactionsDisplaySize": string;
+    /**
+     * リアクションしたユーザーのアバターを表示
+     */
+    "showReactedUserAvatars": string;
     /**
      * リアクションの最大横幅を制限し、縮小して表示する
      */
@@ -6957,6 +7029,10 @@ export interface Locale extends ILocale {
              * ドライブ容量
              */
             "driveCapacity": string;
+            /**
+             * ドライブアップロード帯域
+             */
+            "driveUploadBandwidth": string;
             /**
              * ファイルにNSFWを常に付与
              */
@@ -7075,6 +7151,18 @@ export interface Locale extends ILocale {
              * 「アカウントを見つけやすくする」が有効なユーザー
              */
             "isExplorable": string;
+            /**
+             * メールアドレスが確認済み
+             */
+            "isEmailVerified": string;
+            /**
+             * 2要素認証が有効
+             */
+            "isTwoFactorEnabled": string;
+            /**
+             * パスワードレスログインが有効
+             */
+            "isPasswordLessLoginEnabled": string;
             /**
              * アカウント作成から～以内
              */
@@ -7562,6 +7650,14 @@ export interface Locale extends ILocale {
          * チャンネル外へのリノートと引用リノートを許可する
          */
         "allowRenoteToExternal": string;
+        /**
+         * パブリックの投稿権限が必要
+         */
+        "requirePublicWriteAccess": string;
+        /**
+         * 匿名化戦略
+         */
+        "anonymousStrategy": string;
     };
     "_menuDisplay": {
         /**
@@ -8582,6 +8678,14 @@ export interface Locale extends ILocale {
         "saturday": string;
     };
     "_widgets": {
+        /**
+         * ミリシタイベント情報
+         */
+        "mltdEventInfo": string;
+        /**
+         * ミリシタイベントボーダー
+         */
+        "mltdEventBorder": string;
         /**
          * プロフィール
          */
@@ -8860,6 +8964,10 @@ export interface Locale extends ILocale {
          * チャンネルに投稿...
          */
         "channelPlaceholder": string;
+        /**
+         * {x}に投稿...
+         */
+        "placeholder": ParameterizedString<"x">;
         "_placeholders": {
             /**
              * いまどうしてる？
@@ -10454,6 +10562,10 @@ export interface Locale extends ILocale {
          * 石をアイコンにする
          */
         "useAvatarAsStone": string;
+        /**
+         * 石を10秒ありさにする
+         */
+        "use10secarisaAsStone": string;
     };
     "_offlineScreen": {
         /**
diff --git a/locales/ja-JP.yml b/locales/ja-JP.yml
index 8c803f1ebe..1ce477680e 100644
--- a/locales/ja-JP.yml
+++ b/locales/ja-JP.yml
@@ -20,6 +20,9 @@ cancel: "キャンセル"
 noThankYou: "やめておく"
 enterUsername: "ユーザー名を入力"
 renotedBy: "{user}がリノート"
+renotedByChannel: "{user}が{channel}にリノート"
+publicLetter: "パブリックレター"
+anonymouslySendToUser: "{user}宛のパブリックレター"
 noNotes: "ノートはありません"
 noNotifications: "通知はありません"
 instance: "サーバー"
@@ -114,14 +117,13 @@ enterEmoji: "絵文字を入力"
 renote: "リノート"
 unrenote: "リノート解除"
 renoted: "リノートしました。"
-renotedToX: "{name} にリノートしました。"
 cantRenote: "この投稿はリノートできません。"
 cantReRenote: "リノートをリノートすることはできません。"
 quote: "引用"
 inChannelRenote: "チャンネル内リノート"
 inChannelQuote: "チャンネル内引用"
-renoteToChannel: "チャンネルにリノート"
-renoteToOtherChannel: "他のチャンネルにリノート"
+toChannelRenote: "チャンネルにRenote"
+toChannelQuote: "チャンネルに引用"
 pinnedNote: "ピン留めされたノート"
 pinned: "ピン留め"
 you: "あなた"
@@ -138,6 +140,7 @@ overwriteFromPinnedEmojisForReaction: "リアクション設定から上書き
 overwriteFromPinnedEmojis: "全般設定から上書きする"
 reactionSettingDescription2: "ドラッグして並び替え、クリックして削除、＋を押して追加します。"
 rememberNoteVisibility: "公開範囲を記憶する"
+showNoteVisibilityDial: "公開範囲ダイアルを表示"
 attachCancel: "添付取り消し"
 deleteFile: "ファイルを削除"
 markAsSensitive: "センシティブとして設定"
@@ -376,6 +379,12 @@ today: "今日"
 dayX: "{day}日"
 monthX: "{month}月"
 yearX: "{year}年"
+daily: "日次"
+weekly: "週次"
+monthly: "月次"
+yearly: "年次"
+manual: "手動"
+reactionsAreNotAnonymized: "リアクションは匿名化されません"
 pages: "ページ"
 integration: "連携"
 connectService: "接続する"
@@ -576,6 +585,8 @@ deleteAll: "全て削除"
 showFixedPostForm: "タイムライン上部に投稿フォームを表示する"
 showFixedPostFormInChannel: "タイムライン上部に投稿フォームを表示する(チャンネル)"
 withRepliesByDefaultForNewlyFollowed: "フォローする際、デフォルトで返信をTLに含むようにする"
+postToTheChannelYouHaveOpen: "開いているチャンネルに投稿する"
+autoOpenCws: "自動で開くCW"
 newNoteRecived: "新しいノートがあります"
 sounds: "サウンド"
 sound: "サウンド"
@@ -881,12 +892,16 @@ high: "高"
 middle: "中"
 low: "低"
 emailNotConfiguredWarning: "メールアドレスの設定がされていません。"
+passkeyNotConfiguredWarning: "セキュリティキー・パスキーの設定がされていません。このメッセージが表示されている場合、おそらくあなたが今使っているデバイスはパスキーとして使用できます。"
+twoFactorNotConfiguredWarning: "二要素認証の設定がされていません。"
 ratio: "比率"
 previewNoteText: "本文をプレビュー"
 customCss: "カスタムCSS"
 customCssWarn: "この設定は必ず知識のある方が行ってください。不適切な設定を行うとクライアントが正常に使用できなくなる恐れがあります。"
 global: "グローバル"
 squareAvatars: "アイコンを四角形で表示"
+localNeighborNotes: "ローカルの近隣ノート"
+globalNeighborNotes: "グローバルの近隣ノート"
 sent: "送信"
 received: "受信"
 searchResult: "検索結果"
@@ -1098,7 +1113,10 @@ enableChartsForRemoteUser: "リモートユーザーのチャートを生成"
 enableChartsForFederatedInstances: "リモートサーバーのチャートを生成"
 enableStatsForFederatedInstances: "リモートサーバーの情報を取得"
 showClipButtonInNoteFooter: "ノートのアクションにクリップを追加"
+showTranslateButtonInNoteFooter: "ノートのアクションに翻訳を追加"
 reactionsDisplaySize: "リアクションの表示サイズ"
+adaptiveReactionsDisplaySize: "適応型サイズ"
+showReactedUserAvatars: "リアクションしたユーザーのアバターを表示"
 limitWidthOfReaction: "リアクションの最大横幅を制限し、縮小して表示する"
 noteIdOrUrl: "ノートIDまたはURL"
 video: "動画"
@@ -1798,6 +1816,7 @@ _role:
     canManageCustomEmojis: "カスタム絵文字の管理"
     canManageAvatarDecorations: "アバターデコレーションの管理"
     driveCapacity: "ドライブ容量"
+    driveUploadBandwidth: "ドライブアップロード帯域"
     alwaysMarkNsfw: "ファイルにNSFWを常に付与"
     canUpdateBioMedia: "アイコンとバナーの更新を許可"
     pinMax: "ノートのピン留めの最大数"
@@ -1828,6 +1847,9 @@ _role:
     isSuspended: "サスペンド済みユーザー"
     isLocked: "鍵アカウントユーザー"
     isExplorable: "「アカウントを見つけやすくする」が有効なユーザー"
+    isEmailVerified: "メールアドレスが確認済み"
+    isTwoFactorEnabled: "2要素認証が有効"
+    isPasswordLessLoginEnabled: "パスワードレスログインが有効"
     createdLessThan: "アカウント作成から～以内"
     createdMoreThan: "アカウント作成から～経過"
     followersLessThanOrEq: "フォロワー数が～以下"
@@ -1976,6 +1998,8 @@ _channel:
   nameAndDescription: "名前と説明"
   nameOnly: "名前のみ"
   allowRenoteToExternal: "チャンネル外へのリノートと引用リノートを許可する"
+  requirePublicWriteAccess: "パブリックの投稿権限が必要"
+  anonymousStrategy: "匿名化戦略"
 
 _menuDisplay:
   sideFull: "横"
@@ -2255,6 +2279,8 @@ _weekday:
   saturday: "土曜日"
 
 _widgets:
+  mltdEventInfo: "ミリシタイベント情報"
+  mltdEventBorder: "ミリシタイベントボーダー"
   profile: "プロフィール"
   instanceInfo: "サーバー情報"
   memo: "付箋"
@@ -2331,6 +2357,7 @@ _postForm:
   replyPlaceholder: "このノートに返信..."
   quotePlaceholder: "このノートを引用..."
   channelPlaceholder: "チャンネルに投稿..."
+  placeholder: "{x}に投稿..."
   _placeholders:
     a: "いまどうしてる？"
     b: "何かありましたか？"
@@ -2779,6 +2806,7 @@ _reversi:
   disallowIrregularRules: "変則なし"
   showBoardLabels: "盤面に行・列番号を表示"
   useAvatarAsStone: "石をアイコンにする"
+  use10secarisaAsStone: "石を10秒ありさにする"
 
 _offlineScreen:
   title: "オフライン - サーバーに接続できません"
diff --git a/locales/ja-KS.yml b/locales/ja-KS.yml
index 66560f524b..fb204bd937 100644
--- a/locales/ja-KS.yml
+++ b/locales/ja-KS.yml
@@ -19,6 +19,7 @@ cancel: "やめる"
 noThankYou: "やめとく"
 enterUsername: "ユーザー名を入れてや"
 renotedBy: "{user}がリノートしたで"
+renotedByChannel: "{user}が{channel}にリノートしたで"
 noNotes: "ノートはあらへん"
 noNotifications: "通知はあらへん"
 instance: "サーバー"
@@ -136,6 +137,7 @@ overwriteFromPinnedEmojisForReaction: "リアクション設定から上書き
 overwriteFromPinnedEmojis: "全般設定から上書きする"
 reactionSettingDescription2: "ドラッグで並び替え、クリックで削除、＋を押して追加やで。"
 rememberNoteVisibility: "公開範囲覚えといて"
+showNoteVisibilityDial: "公開範囲ダイアルを表示"
 attachCancel: "のっけるのやめる"
 deleteFile: "ファイルをほかす"
 markAsSensitive: "ちょっと見せられへんわ"
@@ -2271,6 +2273,7 @@ _postForm:
   replyPlaceholder: "このノートに返信..."
   quotePlaceholder: "このノートを引用..."
   channelPlaceholder: "チャンネルに投稿..."
+  placeholder: "{x}に投稿..."
   _placeholders:
     a: "いまどないしとるん？"
     b: "何かあったん？"
diff --git a/locales/ko-KR.yml b/locales/ko-KR.yml
index 36b818c117..85cc46c41e 100644
--- a/locales/ko-KR.yml
+++ b/locales/ko-KR.yml
@@ -20,6 +20,9 @@ cancel: "취소"
 noThankYou: "나중에"
 enterUsername: "유저명 입력"
 renotedBy: "{user}님이 리노트"
+renotedByChannel: "{user}님이 {channel} 채널에서 리노트"
+publicLetter: "공개 서한"
+anonymouslySendToUser: "{user}에게 보내는 공개 서한"
 noNotes: "노트가 없습니다"
 noNotifications: "표시할 알림이 없습니다"
 instance: "서버"
@@ -120,8 +123,8 @@ cantReRenote: "리노트를 리노트 할 수 없습니다."
 quote: "인용"
 inChannelRenote: "채널 내 리노트"
 inChannelQuote: "채널 내 인용"
-renoteToChannel: "채널에 리노트"
-renoteToOtherChannel: "다른 채널에 리노트"
+toChannelRenote: "채널에 리노트"
+toChannelQuote: "채널에 인용"
 pinnedNote: "고정된 노트"
 pinned: "고정하기"
 you: "나"
@@ -138,6 +141,7 @@ overwriteFromPinnedEmojisForReaction: "리액션 설정을 덮어쓰기"
 overwriteFromPinnedEmojis: "일반 설정을 덮어쓰기"
 reactionSettingDescription2: "끌어서 순서 변경, 클릭해서 삭제, ＋를 눌러서 추가할 수 있습니다."
 rememberNoteVisibility: "공개 범위를 기억하기"
+showNoteVisibilityDial: "노트 공개 범위 다이얼 표시"
 attachCancel: "첨부 취소"
 deleteFile: "파일 삭제"
 markAsSensitive: "열람주의로 설정"
@@ -376,6 +380,12 @@ today: "오늘"
 dayX: "{day}일"
 monthX: "{month}월"
 yearX: "{year}년"
+daily: "일일"
+weekly: "주간"
+monthly: "월간"
+yearly: "연간"
+manual: "수동"
+reactionsAreNotAnonymized: "리액션은 익명화되지 않습니다"
 pages: "페이지"
 integration: "연동"
 connectService: "계정 연동"
@@ -576,6 +586,8 @@ deleteAll: "모두 삭제"
 showFixedPostForm: "타임라인 상단에 글 작성란을 표시"
 showFixedPostFormInChannel: "채널 타임라인 상단에 글 작성란을 표시"
 withRepliesByDefaultForNewlyFollowed: "팔로우 할 때 기본적으로 답글을 타임라인에 나오게 하기"
+postToTheChannelYouHaveOpen: "현재 보고 있는 채널에 글 작성"
+autoOpenCws: "자동으로 열어야 할 CW"
 newNoteRecived: "새 노트가 있습니다"
 sounds: "소리"
 sound: "소리"
@@ -881,12 +893,16 @@ high: "높음"
 middle: "보통"
 low: "낮음"
 emailNotConfiguredWarning: "메일 주소가 설정되어 있지 않습니다."
+passkeyNotConfiguredWarning: "패스 키가 설정되어 있지 않습니다. 이 메시지를 보면 사용 중인 장치가 보안 키로 사용될 수 있습니다."
+twoFactorNotConfiguredWarning: "2단계 인증이 설정되어 있지 않습니다."
 ratio: "비율"
 previewNoteText: "본문 미리보기"
 customCss: "CSS 사용자화"
 customCssWarn: "이 설정은 기능을 알고 있는 경우에만 사용해야 합니다. 잘못된 값을 입력하면 클라이언트가 정상적으로 작동하지 않을 수 있습니다."
 global: "글로벌"
 squareAvatars: "프로필 아바타를 사각형으로 표시"
+localNeighborNotes: "이 서버의 이웃 노트"
+globalNeighborNotes: "글로벌 이웃 노트"
 sent: "전송"
 received: "수신"
 searchResult: "검색 결과"
@@ -1098,8 +1114,11 @@ enableChartsForRemoteUser: "리모트 유저의 차트를 생성"
 enableChartsForFederatedInstances: "리모트 서버의 차트를 생성"
 enableStatsForFederatedInstances: "리모트 서버 정보 받아오기"
 showClipButtonInNoteFooter: "노트 동작에 클립을 추가"
+showTranslateButtonInNoteFooter: "노트 동작에 번역을 추가"
 reactionsDisplaySize: "리액션 표시 크기"
+adaptiveReactionsDisplaySize: "적응형 크기"
 limitWidthOfReaction: "리액션의 최대 폭을 제한하고 작게 표시하기"
+showReactedUserAvatars: "리액션한 유저의 아바타 표시"
 noteIdOrUrl: "노트 ID 및 URL"
 video: "동영상"
 videos: "동영상"
@@ -1783,6 +1802,7 @@ _role:
     canManageCustomEmojis: "커스텀 이모지 관리"
     canManageAvatarDecorations: "아바타 꾸미기 관리"
     driveCapacity: "드라이브 용량"
+    driveUploadBandwidth: "드라이브 업로드 대역폭"
     alwaysMarkNsfw: "파일을 항상 NSFW로 지정"
     canUpdateBioMedia: "아바타 및 배너 이미지 변경 허용"
     pinMax: "고정할 수 있는 노트 수"
@@ -1813,6 +1833,9 @@ _role:
     isSuspended: "정지된 사용자"
     isLocked: "잠금 계정 사용자"
     isExplorable: "‘계정을 쉽게 발견하도록 하기’를 활성화한 사용자"
+    isEmailVerified: "이메일 인증 완료"
+    isTwoFactorEnabled: "2단계 인증 활성화"
+    isPasswordLessLoginEnabled: "비밀번호 없는 로그인 활성화"
     createdLessThan: "가입한 지 다음 일수 이내인 유저"
     createdMoreThan: "가입한 지 다음 일수 이상인 유저"
     followersLessThanOrEq: "팔로워 수가 다음 이하인 유저"
@@ -1944,6 +1967,8 @@ _channel:
   nameAndDescription: "이름과 설명"
   nameOnly: "이름만"
   allowRenoteToExternal: "채널 외부로의 리노트와 인용 리노트를 허가"
+  requirePublicWriteAccess: "노트 작성 권한을 공개로 제한"
+  anonymousStrategy: "익명 전략"
 _menuDisplay:
   sideFull: "가로"
   sideIcon: "가로(아이콘)"
@@ -2279,6 +2304,7 @@ _postForm:
   replyPlaceholder: "이 노트에 답글..."
   quotePlaceholder: "이 노트를 인용..."
   channelPlaceholder: "채널에 게시하기..."
+  placeholder: "{x}에 게시하기..."
   _placeholders:
     a: "지금 무엇을 하고 있나요?"
     b: "무슨 일이 일어나고 있나요?"
diff --git a/locales/zh-CN.yml b/locales/zh-CN.yml
index f4df425af4..f97b67e91e 100644
--- a/locales/zh-CN.yml
+++ b/locales/zh-CN.yml
@@ -20,6 +20,9 @@ cancel: "取消"
 noThankYou: "不用，谢谢"
 enterUsername: "输入用户名"
 renotedBy: "{user} 转发了"
+renotedByChannel: "{user} 转发了 {channel}"
+publicLetter: "公开信"
+anonymouslySendToUser: "至{user}的公开信"
 noNotes: "没有帖文"
 noNotifications: "无通知"
 instance: "服务器"
@@ -120,8 +123,8 @@ cantReRenote: "转发无法被再次转发。"
 quote: "引用"
 inChannelRenote: "在频道内转发"
 inChannelQuote: "在频道内引用"
-renoteToChannel: "转帖至频道"
-renoteToOtherChannel: "转帖至其它频道"
+toChannelRenote: "转发到频道"
+toChannelQuote: "引用到频道"
 pinnedNote: "已置顶的帖子"
 pinned: "置顶"
 you: "您"
@@ -138,6 +141,7 @@ overwriteFromPinnedEmojisForReaction: "从「置顶（回应）」设置覆盖"
 overwriteFromPinnedEmojis: "从全局设置覆盖"
 reactionSettingDescription2: "拖动重新排序，单击删除，点击 + 添加。"
 rememberNoteVisibility: "保存上次设置的可见性"
+showNoteVisibilityDial: "显示可见性拨号"
 attachCancel: "取消添加附件"
 deleteFile: "删除文件"
 markAsSensitive: "标记为敏感内容"
@@ -376,6 +380,12 @@ today: "今天"
 dayX: "{day}日"
 monthX: "{month}月"
 yearX: "{year}年"
+daily: "每日"
+weekly: "每周"
+monthly: "每月"
+yearly: "每年"
+manual: "手动"
+reactionsAreNotAnonymized: "回应不是匿名的"
 pages: "页面"
 integration: "关联"
 connectService: "连接"
@@ -576,6 +586,8 @@ deleteAll: "全部删除"
 showFixedPostForm: "在时间线顶部显示发帖框"
 showFixedPostFormInChannel: "在时间线顶部显示发帖对话框（频道）"
 withRepliesByDefaultForNewlyFollowed: "在时间线中默认包含新关注用户的回复"
+postToTheChannelYouHaveOpen: "在当前打开的频道中发帖"
+autoOpenCws: "应自动打开的内容警告"
 newNoteRecived: "有新的帖子"
 sounds: "提示音"
 sound: "提示音"
@@ -881,12 +893,16 @@ high: "高"
 middle: "中"
 low: "低"
 emailNotConfiguredWarning: "尚未设置电子邮件地址。"
+passkeyNotConfiguredWarning: "Passkey 未设置。如果您看到这条消息，您正在使用的设备很可能可以作为安全密钥。"
+twoFactorNotConfiguredWarning: "双因素认证未设置。"
 ratio: "比率"
 previewNoteText: "预览文本"
 customCss: "自定义 CSS"
 customCssWarn: "这些设置必须有相关的基础知识，不当的配置可能导致客户端无法正常使用。"
 global: "全局"
 squareAvatars: "显示方形头像图标"
+localNeighborNotes: "本地邻居帖子"
+globalNeighborNotes: "全局邻居帖子"
 sent: "发送"
 received: "收取"
 searchResult: "搜索结果"
@@ -1098,8 +1114,11 @@ enableChartsForRemoteUser: "生成远程用户的图表"
 enableChartsForFederatedInstances: "生成远程服务器的图表"
 enableStatsForFederatedInstances: "获取远程服务器的信息"
 showClipButtonInNoteFooter: "在贴文下方显示便签按钮"
+showTranslateButtonInNoteFooter: "在贴文下方显示翻译按钮"
 reactionsDisplaySize: "回应显示大小"
+adaptiveReactionsDisplaySize: "自适应大小"
 limitWidthOfReaction: "限制回应的最大宽度，并将其缩小显示"
+showReactedUserAvatars: "显示回应者的头像"
 noteIdOrUrl: "帖子 ID 或 URL"
 video: "视频"
 videos: "视频"
@@ -1783,6 +1802,7 @@ _role:
     canManageCustomEmojis: "管理自定义表情符号"
     canManageAvatarDecorations: "管理头像挂件"
     driveCapacity: "网盘容量"
+    driveUploadBandwidth: "网盘上传带宽"
     alwaysMarkNsfw: "总是将文件标记为 NSFW"
     canUpdateBioMedia: "可以更新头像和横幅"
     pinMax: "帖子置顶数量限制"
@@ -1813,6 +1833,9 @@ _role:
     isSuspended: "停用的用户"
     isLocked: "锁推用户"
     isExplorable: "启用“使账号可见”的用户"
+    isEmailVerified: "电子邮件已验证"
+    isTwoFactorEnabled: "已启用两步验证"
+    isPasswordLessLoginEnabled: "已启用无密码登录"
     createdLessThan: "账户创建时间少于"
     createdMoreThan: "账户创建时间超过"
     followersLessThanOrEq: "关注者不多于"
@@ -1944,6 +1967,8 @@ _channel:
   nameAndDescription: "名称与描述"
   nameOnly: "仅名称"
   allowRenoteToExternal: "允许在频道外转帖及引用"
+  requirePublicWriteAccess: "需要公开写入权限"
+  anonymousStrategy: "匿名策略"
 _menuDisplay:
   sideFull: "横向"
   sideIcon: "横向(图标)"
@@ -2278,7 +2303,8 @@ _visibility:
 _postForm:
   replyPlaceholder: "回复这个帖子..."
   quotePlaceholder: "引用这个帖子..."
-  channelPlaceholder: "发布到频道…"
+  channelPlaceholder: "发布到频道..."
+  placeholder: "发布到{x}..."
   _placeholders:
     a: "现在如何？"
     b: "发生了什么？"
diff --git a/locales/zh-TW.yml b/locales/zh-TW.yml
index 466e3cc1d8..cdc258b918 100644
--- a/locales/zh-TW.yml
+++ b/locales/zh-TW.yml
@@ -20,6 +20,9 @@ cancel: "取消"
 noThankYou: "現在不要"
 enterUsername: "輸入使用者名稱"
 renotedBy: "{user} 轉發了"
+renotedByChannel: "{user} 轉發了 {channel}"
+publicLetter: "公開信"
+anonymouslySendToUser: "至{user}的公開信"
 noNotes: "無貼文"
 noNotifications: "沒有通知"
 instance: "伺服器"
@@ -120,8 +123,8 @@ cantReRenote: "無法轉發之前已經轉發過的內容。"
 quote: "引用"
 inChannelRenote: "在頻道內轉發"
 inChannelQuote: "在頻道內引用"
-renoteToChannel: "轉發至頻道"
-renoteToOtherChannel: "轉發至其他頻道"
+toChannelRenote: "轉發到頻道"
+toChannelQuote: "引用到頻道"
 pinnedNote: "已置頂的貼文"
 pinned: "置頂"
 you: "您"
@@ -138,6 +141,7 @@ overwriteFromPinnedEmojisForReaction: "從反應複寫設定"
 overwriteFromPinnedEmojis: "從一般複寫設定"
 reactionSettingDescription2: "拖動以交換，點擊以刪除，按下「+」以新增。"
 rememberNoteVisibility: "記住貼文可見性"
+showNoteVisibilityDial: "顯示可視性撥號"
 attachCancel: "移除附件"
 deleteFile: "刪除檔案"
 markAsSensitive: "標記為敏感內容"
@@ -376,6 +380,12 @@ today: "本日"
 dayX: "{day} 日"
 monthX: "{month} 月"
 yearX: "{year} 年"
+daily: "每日"
+weekly: "每週"
+monthly: "每月"
+yearly: "每年"
+manual: "手動"
+reactionsAreNotAnonymized: "反應不是匿名的"
 pages: "頁面"
 integration: "整合"
 connectService: "已連結"
@@ -576,6 +586,8 @@ deleteAll: "刪除所有記錄"
 showFixedPostForm: "於時間軸頁頂顯示「發送貼文」方框"
 showFixedPostFormInChannel: "於時間軸頁頂顯示「發送貼文」方框（頻道）"
 withRepliesByDefaultForNewlyFollowed: "在追隨其他人後，預設在時間軸納入回覆的貼文"
+postToTheChannelYouHaveOpen: "發送至您正在瀏覽的頻道"
+autoOpenCws: "應自動展開的警告"
 newNoteRecived: "發現新貼文"
 sounds: "音效"
 sound: "音效"
@@ -881,12 +893,16 @@ high: "高"
 middle: "中"
 low: "低"
 emailNotConfiguredWarning: "沒有設定電子郵件地址"
+passkeyNotConfiguredWarning: "沒有設定 Passkey。如果您看到這條消息，您正在使用的設備很可能可以作為安全金鑰。"
+twoFactorNotConfiguredWarning: "沒有設定雙重驗證"
 ratio: "%"
 previewNoteText: "預覽文本"
 customCss: "自定義 CSS"
 customCssWarn: "這個設定必須由具備相關知識的人員操作，不當的設定可能導致客戶端無法正常使用。"
 global: "全域"
 squareAvatars: "大頭貼以方形顯示"
+localNeighborNotes: "本地鄰近貼文"
+globalNeighborNotes: "全域鄰近貼文"
 sent: "發送"
 received: "收取"
 searchResult: "搜尋結果"
@@ -1098,8 +1114,11 @@ enableChartsForRemoteUser: "生成遠端使用者的圖表"
 enableChartsForFederatedInstances: "生成遠端伺服器的圖表"
 enableStatsForFederatedInstances: "取得遠端伺服器資訊"
 showClipButtonInNoteFooter: "新增摘錄按鈕至貼文"
+showTranslateButtonInNoteFooter: "新增翻譯按鈕至貼文"
 reactionsDisplaySize: "反應的顯示尺寸"
+adaptiveReactionsDisplaySize: "自適應尺寸"
 limitWidthOfReaction: "限制反應的最大寬度，並縮小顯示尺寸。"
+showReactedUserAvatars: "顯示反應的使用者頭像"
 noteIdOrUrl: "貼文 ID 或 URL"
 video: "影片"
 videos: "影片"
@@ -1783,6 +1802,7 @@ _role:
     canManageCustomEmojis: "管理自訂表情符號"
     canManageAvatarDecorations: "管理頭像裝飾"
     driveCapacity: "雲端硬碟容量"
+    driveUploadBandwidth: "雲端硬碟上傳頻寬"
     alwaysMarkNsfw: "總是將檔案標記為NSFW"
     canUpdateBioMedia: "允許更新大頭貼和橫幅"
     pinMax: "置頂貼文的最大數量"
@@ -1944,6 +1964,8 @@ _channel:
   nameAndDescription: "名稱"
   nameOnly: "僅名稱"
   allowRenoteToExternal: "允許在頻道外轉發和引用"
+  requirePublicWriteAccess: "需要公開寫入權限"
+  anonymousStrategy: "匿名策略"
 _menuDisplay:
   sideFull: "橫向"
   sideIcon: "橫向（圖示）"
@@ -2278,7 +2300,8 @@ _visibility:
 _postForm:
   replyPlaceholder: "回覆此貼文..."
   quotePlaceholder: "引用此貼文..."
-  channelPlaceholder: "發佈到頻道"
+  channelPlaceholder: "發佈到頻道..."
+  placeholder: "發佈到{x}..."
   _placeholders:
     a: "今天過得如何？"
     b: "有什麼新鮮事嗎？"
diff --git a/package.json b/package.json
index c6e20945a1..8b54d53944 100644
--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
 	"name": "misskey",
-	"version": "2025.2.1-beta.0",
+	"version": "2025.2.1-ltn.26626",
 	"codename": "nasubi",
 	"repository": {
 		"type": "git",
diff --git a/packages/backend/package.json b/packages/backend/package.json
index cee5c7205b..4c188ecbcc 100644
--- a/packages/backend/package.json
+++ b/packages/backend/package.json
@@ -145,6 +145,7 @@
 		"nanoid": "5.1.0",
 		"nested-property": "4.0.0",
 		"node-fetch": "3.3.2",
+		"node-sql-parser": "5.3.6",
 		"nodemailer": "6.10.0",
 		"nsfwjs": "4.2.0",
 		"oauth": "0.10.0",
@@ -154,6 +155,7 @@
 		"otpauth": "9.3.6",
 		"parse5": "7.2.1",
 		"pg": "8.13.3",
+		"pg-query-stream": "4.7.3",
 		"pkce-challenge": "4.1.0",
 		"probe-image-size": "7.2.3",
 		"promise-limit": "2.7.0",
diff --git a/packages/backend/src/di-symbols.ts b/packages/backend/src/di-symbols.ts
index e599fc7b37..d1c1e7d453 100644
--- a/packages/backend/src/di-symbols.ts
+++ b/packages/backend/src/di-symbols.ts
@@ -6,6 +6,7 @@
 export const DI = {
 	config: Symbol('config'),
 	db: Symbol('db'),
+	dbLong: Symbol('dbLong'),
 	meta: Symbol('meta'),
 	meilisearch: Symbol('meilisearch'),
 	redis: Symbol('redis'),
@@ -69,6 +70,7 @@ export const DI = {
 	promoReadsRepository: Symbol('promoReadsRepository'),
 	relaysRepository: Symbol('relaysRepository'),
 	channelsRepository: Symbol('channelsRepository'),
+	channelAnonymousSaltsRepository: Symbol('channelAnonymousSaltsRepository'),
 	channelFollowingsRepository: Symbol('channelFollowingsRepository'),
 	channelFavoritesRepository: Symbol('channelFavoritesRepository'),
 	registryItemsRepository: Symbol('registryItemsRepository'),
diff --git a/packages/backend/src/models/RepositoryModule.ts b/packages/backend/src/models/RepositoryModule.ts
index ea0f88baba..7479e8175a 100644
--- a/packages/backend/src/models/RepositoryModule.ts
+++ b/packages/backend/src/models/RepositoryModule.ts
@@ -20,6 +20,7 @@ import {
 	MiBlocking,
 	MiBubbleGameRecord,
 	MiChannel,
+	MiChannelAnonymousSalt,
 	MiChannelFavorite,
 	MiChannelFollowing,
 	MiClip,
@@ -405,6 +406,12 @@ const $channelsRepository: Provider = {
 	inject: [DI.db],
 };
 
+const $channelAnonymousSaltsRepository: Provider = {
+	provide: DI.channelAnonymousSaltsRepository,
+	useFactory: (db: DataSource) => db.getRepository(MiChannelAnonymousSalt),
+	inject: [DI.db],
+};
+
 const $channelFollowingsRepository: Provider = {
 	provide: DI.channelFollowingsRepository,
 	useFactory: (db: DataSource) => db.getRepository(MiChannelFollowing).extend(miRepository as MiRepository<MiChannelFollowing>),
@@ -552,6 +559,7 @@ const $reversiGamesRepository: Provider = {
 		$promoReadsRepository,
 		$relaysRepository,
 		$channelsRepository,
+		$channelAnonymousSaltsRepository,
 		$channelFollowingsRepository,
 		$channelFavoritesRepository,
 		$registryItemsRepository,
@@ -623,6 +631,7 @@ const $reversiGamesRepository: Provider = {
 		$promoReadsRepository,
 		$relaysRepository,
 		$channelsRepository,
+		$channelAnonymousSaltsRepository,
 		$channelFollowingsRepository,
 		$channelFavoritesRepository,
 		$registryItemsRepository,
diff --git a/packages/backend/src/models/_.ts b/packages/backend/src/models/_.ts
index c72bdaa727..6db277c58d 100644
--- a/packages/backend/src/models/_.ts
+++ b/packages/backend/src/models/_.ts
@@ -21,6 +21,7 @@ import { MiApp } from '@/models/App.js';
 import { MiAvatarDecoration } from '@/models/AvatarDecoration.js';
 import { MiAuthSession } from '@/models/AuthSession.js';
 import { MiBlocking } from '@/models/Blocking.js';
+import { MiChannelAnonymousSalt } from '@/models/ChannelAnonymousSalt.js';
 import { MiChannelFollowing } from '@/models/ChannelFollowing.js';
 import { MiChannelFavorite } from '@/models/ChannelFavorite.js';
 import { MiClip } from '@/models/Clip.js';
@@ -136,6 +137,7 @@ export {
 	MiAvatarDecoration,
 	MiAuthSession,
 	MiBlocking,
+	MiChannelAnonymousSalt,
 	MiChannelFollowing,
 	MiChannelFavorite,
 	MiClip,
@@ -207,6 +209,7 @@ export type AppsRepository = Repository<MiApp> & MiRepository<MiApp>;
 export type AvatarDecorationsRepository = Repository<MiAvatarDecoration> & MiRepository<MiAvatarDecoration>;
 export type AuthSessionsRepository = Repository<MiAuthSession> & MiRepository<MiAuthSession>;
 export type BlockingsRepository = Repository<MiBlocking> & MiRepository<MiBlocking>;
+export type ChannelAnonymousSaltsRepository = Repository<MiChannelAnonymousSalt> & MiRepository<MiChannelAnonymousSalt>;
 export type ChannelFollowingsRepository = Repository<MiChannelFollowing> & MiRepository<MiChannelFollowing>;
 export type ChannelFavoritesRepository = Repository<MiChannelFavorite> & MiRepository<MiChannelFavorite>;
 export type ClipsRepository = Repository<MiClip> & MiRepository<MiClip>;
diff --git a/packages/backend/src/server/api/ApiServerService.ts b/packages/backend/src/server/api/ApiServerService.ts
index 3a8cb19f01..e81b801757 100644
--- a/packages/backend/src/server/api/ApiServerService.ts
+++ b/packages/backend/src/server/api/ApiServerService.ts
@@ -11,11 +11,13 @@ import { ModuleRef } from '@nestjs/core';
 import { AuthenticationResponseJSON } from '@simplewebauthn/types';
 import type { Config } from '@/config.js';
 import type { InstancesRepository, AccessTokensRepository } from '@/models/_.js';
+import type Logger from '@/logger.js';
 import { DI } from '@/di-symbols.js';
 import { UserEntityService } from '@/core/entities/UserEntityService.js';
 import { bindThis } from '@/decorators.js';
 import endpoints from './endpoints.js';
 import { ApiCallService } from './ApiCallService.js';
+import { ApiLoggerService } from './ApiLoggerService.js';
 import { SignupApiService } from './SignupApiService.js';
 import { SigninApiService } from './SigninApiService.js';
 import { SigninWithPasskeyApiService } from './SigninWithPasskeyApiService.js';
@@ -23,6 +25,8 @@ import type { FastifyInstance, FastifyPluginOptions } from 'fastify';
 
 @Injectable()
 export class ApiServerService {
+	private logger: Logger;
+
 	constructor(
 		private moduleRef: ModuleRef,
 
@@ -37,10 +41,12 @@ export class ApiServerService {
 
 		private userEntityService: UserEntityService,
 		private apiCallService: ApiCallService,
+		private apiLoggerService: ApiLoggerService,
 		private signupApiService: SignupApiService,
 		private signinApiService: SigninApiService,
 		private signinWithPasskeyApiService: SigninWithPasskeyApiService,
 	) {
+		this.logger = this.apiLoggerService.logger;
 		//this.createServer = this.createServer.bind(this);
 	}
 
@@ -121,7 +127,10 @@ export class ApiServerService {
 				'm-captcha-response'?: string;
 				'testcaptcha-response'?: string;
 			}
-		}>('/signup', (request, reply) => this.signupApiService.signup(request, reply));
+		}>('/signup', (request, reply) => this.signupApiService.signup(request, reply).catch(err => {
+			this.logger.error(err);
+			throw err;
+		}));
 
 		fastify.post<{
 			Body: {
@@ -135,14 +144,20 @@ export class ApiServerService {
 				'm-captcha-response'?: string;
 				'testcaptcha-response'?: string;
 			};
-		}>('/signin-flow', (request, reply) => this.signinApiService.signin(request, reply));
+		}>('/signin-flow', (request, reply) => this.signinApiService.signin(request, reply).catch(err => {
+			this.logger.error(err);
+			throw err;
+		}));
 
 		fastify.post<{
 			Body: {
 				credential?: AuthenticationResponseJSON;
 				context?: string;
 			};
-		}>('/signin-with-passkey', (request, reply) => this.signinWithPasskeyApiService.signin(request, reply));
+		}>('/signin-with-passkey', (request, reply) => this.signinWithPasskeyApiService.signin(request, reply).catch(err => {
+			this.logger.error(err);
+			throw err;
+		}));
 
 		fastify.post<{ Body: { code: string; } }>('/signup-pending', (request, reply) => this.signupApiService.signupPending(request, reply));
 
diff --git a/packages/backend/src/server/api/endpoint-list.ts b/packages/backend/src/server/api/endpoint-list.ts
index 28f7cfea04..d107e9e979 100644
--- a/packages/backend/src/server/api/endpoint-list.ts
+++ b/packages/backend/src/server/api/endpoint-list.ts
@@ -256,6 +256,7 @@ export * as 'i/import-blocking' from './endpoints/i/import-blocking.js';
 export * as 'i/import-following' from './endpoints/i/import-following.js';
 export * as 'i/import-muting' from './endpoints/i/import-muting.js';
 export * as 'i/import-user-lists' from './endpoints/i/import-user-lists.js';
+export * as 'i/jobs' from './endpoints/i/jobs.js';
 export * as 'i/move' from './endpoints/i/move.js';
 export * as 'i/notifications' from './endpoints/i/notifications.js';
 export * as 'i/notifications-grouped' from './endpoints/i/notifications-grouped.js';
@@ -367,6 +368,7 @@ export * as 'test' from './endpoints/test.js';
 export * as 'username/available' from './endpoints/username/available.js';
 export * as 'users' from './endpoints/users.js';
 export * as 'users/achievements' from './endpoints/users/achievements.js';
+export * as 'users/avatars' from './endpoints/users/avatars.js';
 export * as 'users/clips' from './endpoints/users/clips.js';
 export * as 'users/featured-notes' from './endpoints/users/featured-notes.js';
 export * as 'users/flashs' from './endpoints/users/flashs.js';
diff --git a/packages/frontend/src/scripts/misskey-api.ts b/packages/frontend/src/scripts/misskey-api.ts
index dc07ad477b..22a065d716 100644
--- a/packages/frontend/src/scripts/misskey-api.ts
+++ b/packages/frontend/src/scripts/misskey-api.ts
@@ -82,6 +82,7 @@ export function misskeyApiGet<
 >(
 	endpoint: E,
 	data: P = {} as any,
+	signal?: AbortSignal,
 ): Promise<_ResT> {
 	pendingApiRequestsCount.value++;
 
@@ -89,7 +90,7 @@ export function misskeyApiGet<
 		pendingApiRequestsCount.value--;
 	};
 
-	const query = new URLSearchParams(data as any);
+	const query = new URLSearchParams(Object.entries(data).filter(([, v]) => v !== undefined) as [string, string][]);
 
 	const promise = new Promise<_ResT>((resolve, reject) => {
 		// Send request
@@ -97,6 +98,7 @@ export function misskeyApiGet<
 			method: 'GET',
 			credentials: 'omit',
 			cache: 'default',
+			signal,
 		}).then(async (res) => {
 			const body = res.status === 204 ? null : await res.json();
 
diff --git a/packages/misskey-js/etc/misskey-js.api.md b/packages/misskey-js/etc/misskey-js.api.md
index 1b935364f1..2e4f26ffaa 100644
--- a/packages/misskey-js/etc/misskey-js.api.md
+++ b/packages/misskey-js/etc/misskey-js.api.md
@@ -1600,6 +1600,7 @@ declare namespace entities {
         IImportFollowingRequest,
         IImportMutingRequest,
         IImportUserListsRequest,
+        IJobsResponse,
         IMoveRequest,
         IMoveResponse,
         INotificationsRequest,
@@ -1766,6 +1767,8 @@ declare namespace entities {
         UsersResponse,
         UsersAchievementsRequest,
         UsersAchievementsResponse,
+        UsersAvatarsRequest,
+        UsersAvatarsResponse,
         UsersClipsRequest,
         UsersClipsResponse,
         UsersFeaturedNotesRequest,
@@ -2248,6 +2251,9 @@ type IImportMutingRequest = operations['i___import-muting']['requestBody']['cont
 // @public (undocumented)
 type IImportUserListsRequest = operations['i___import-user-lists']['requestBody']['content']['application/json'];
 
+// @public (undocumented)
+type IJobsResponse = operations['i___jobs']['responses']['200']['content']['application/json'];
+
 // @public (undocumented)
 type IMoveRequest = operations['i___move']['requestBody']['content']['application/json'];
 
@@ -3270,6 +3276,12 @@ type UsersAchievementsRequest = operations['users___achievements']['requestBody'
 // @public (undocumented)
 type UsersAchievementsResponse = operations['users___achievements']['responses']['200']['content']['application/json'];
 
+// @public (undocumented)
+type UsersAvatarsRequest = operations['users___avatars']['requestBody']['content']['application/json'];
+
+// @public (undocumented)
+type UsersAvatarsResponse = operations['users___avatars']['responses']['200']['content']['application/json'];
+
 // @public (undocumented)
 type UsersClipsRequest = operations['users___clips']['requestBody']['content']['application/json'];
 
diff --git a/packages/misskey-js/src/autogen/apiClientJSDoc.ts b/packages/misskey-js/src/autogen/apiClientJSDoc.ts
index 6bace3924c..dd17396bb8 100644
--- a/packages/misskey-js/src/autogen/apiClientJSDoc.ts
+++ b/packages/misskey-js/src/autogen/apiClientJSDoc.ts
@@ -2749,6 +2749,18 @@ declare module '../api.js' {
       credential?: string | null,
     ): Promise<SwitchCaseResponseType<E, P>>;
 
+    /**
+     * No description provided.
+     * 
+     * **Internal Endpoint**: This endpoint is an API for the misskey mainframe and is not intended for use by third parties.
+     * **Credential required**: *Yes*
+     */
+    request<E extends 'i/jobs', P extends Endpoints[E]['req']>(
+      endpoint: E,
+      params: P,
+      credential?: string | null,
+    ): Promise<SwitchCaseResponseType<E, P>>;
+
     /**
      * No description provided.
      * 
@@ -3982,6 +3994,17 @@ declare module '../api.js' {
       credential?: string | null,
     ): Promise<SwitchCaseResponseType<E, P>>;
 
+    /**
+     * List avatars of users.
+     * 
+     * **Credential required**: *No*
+     */
+    request<E extends 'users/avatars', P extends Endpoints[E]['req']>(
+      endpoint: E,
+      params: P,
+      credential?: string | null,
+    ): Promise<SwitchCaseResponseType<E, P>>;
+
     /**
      * Show all clips this user owns.
      * 
diff --git a/packages/misskey-js/src/autogen/endpoint.ts b/packages/misskey-js/src/autogen/endpoint.ts
index 2de0bc9808..f0a7dfa97f 100644
--- a/packages/misskey-js/src/autogen/endpoint.ts
+++ b/packages/misskey-js/src/autogen/endpoint.ts
@@ -365,6 +365,7 @@ import type {
 	IImportFollowingRequest,
 	IImportMutingRequest,
 	IImportUserListsRequest,
+	IJobsResponse,
 	IMoveRequest,
 	IMoveResponse,
 	INotificationsRequest,
@@ -531,6 +532,8 @@ import type {
 	UsersResponse,
 	UsersAchievementsRequest,
 	UsersAchievementsResponse,
+	UsersAvatarsRequest,
+	UsersAvatarsResponse,
 	UsersClipsRequest,
 	UsersClipsResponse,
 	UsersFeaturedNotesRequest,
@@ -832,6 +835,7 @@ export type Endpoints = {
 	'i/import-following': { req: IImportFollowingRequest; res: EmptyResponse };
 	'i/import-muting': { req: IImportMutingRequest; res: EmptyResponse };
 	'i/import-user-lists': { req: IImportUserListsRequest; res: EmptyResponse };
+	'i/jobs': { req: EmptyRequest; res: IJobsResponse };
 	'i/move': { req: IMoveRequest; res: IMoveResponse };
 	'i/notifications': { req: INotificationsRequest; res: INotificationsResponse };
 	'i/notifications-grouped': { req: INotificationsGroupedRequest; res: INotificationsGroupedResponse };
@@ -943,6 +947,7 @@ export type Endpoints = {
 	'username/available': { req: UsernameAvailableRequest; res: UsernameAvailableResponse };
 	'users': { req: UsersRequest; res: UsersResponse };
 	'users/achievements': { req: UsersAchievementsRequest; res: UsersAchievementsResponse };
+	'users/avatars': { req: UsersAvatarsRequest; res: UsersAvatarsResponse };
 	'users/clips': { req: UsersClipsRequest; res: UsersClipsResponse };
 	'users/featured-notes': { req: UsersFeaturedNotesRequest; res: UsersFeaturedNotesResponse };
 	'users/flashs': { req: UsersFlashsRequest; res: UsersFlashsResponse };
diff --git a/packages/misskey-js/src/autogen/entities.ts b/packages/misskey-js/src/autogen/entities.ts
index b7639abca8..2816e3fd7e 100644
--- a/packages/misskey-js/src/autogen/entities.ts
+++ b/packages/misskey-js/src/autogen/entities.ts
@@ -368,6 +368,7 @@ export type IImportBlockingRequest = operations['i___import-blocking']['requestB
 export type IImportFollowingRequest = operations['i___import-following']['requestBody']['content']['application/json'];
 export type IImportMutingRequest = operations['i___import-muting']['requestBody']['content']['application/json'];
 export type IImportUserListsRequest = operations['i___import-user-lists']['requestBody']['content']['application/json'];
+export type IJobsResponse = operations['i___jobs']['responses']['200']['content']['application/json'];
 export type IMoveRequest = operations['i___move']['requestBody']['content']['application/json'];
 export type IMoveResponse = operations['i___move']['responses']['200']['content']['application/json'];
 export type INotificationsRequest = operations['i___notifications']['requestBody']['content']['application/json'];
@@ -534,6 +535,8 @@ export type UsersRequest = operations['users']['requestBody']['content']['applic
 export type UsersResponse = operations['users']['responses']['200']['content']['application/json'];
 export type UsersAchievementsRequest = operations['users___achievements']['requestBody']['content']['application/json'];
 export type UsersAchievementsResponse = operations['users___achievements']['responses']['200']['content']['application/json'];
+export type UsersAvatarsRequest = operations['users___avatars']['requestBody']['content']['application/json'];
+export type UsersAvatarsResponse = operations['users___avatars']['responses']['200']['content']['application/json'];
 export type UsersClipsRequest = operations['users___clips']['requestBody']['content']['application/json'];
 export type UsersClipsResponse = operations['users___clips']['responses']['200']['content']['application/json'];
 export type UsersFeaturedNotesRequest = operations['users___featured-notes']['requestBody']['content']['application/json'];
diff --git a/packages/misskey-js/src/autogen/types.ts b/packages/misskey-js/src/autogen/types.ts
index c7485c6c3d..b57b8a8d9f 100644
--- a/packages/misskey-js/src/autogen/types.ts
+++ b/packages/misskey-js/src/autogen/types.ts
@@ -2406,6 +2406,16 @@ export type paths = {
      */
     post: operations['i___import-user-lists'];
   };
+  '/i/jobs': {
+    /**
+     * i/jobs
+     * @description No description provided.
+     *
+     * **Internal Endpoint**: This endpoint is an API for the misskey mainframe and is not intended for use by third parties.
+     * **Credential required**: *Yes*
+     */
+    post: operations['i___jobs'];
+  };
   '/i/move': {
     /**
      * i/move
@@ -3445,6 +3455,15 @@ export type paths = {
      */
     post: operations['users___achievements'];
   };
+  '/users/avatars': {
+    /**
+     * users/avatars
+     * @description List avatars of users.
+     *
+     * **Credential required**: *No*
+     */
+    post: operations['users___avatars'];
+  };
   '/users/clips': {
     /**
      * users/clips
@@ -4136,6 +4155,13 @@ export type components = {
       renoteId?: string | null;
       reply?: components['schemas']['Note'] | null;
       renote?: components['schemas']['Note'] | null;
+      /**
+       * Format: id
+       * @example xxxxxxxxxx
+       */
+      anonymouslySendToUserId?: string | null;
+      anonymouslySendToUser?: components['schemas']['UserLite'] | null;
+      anonymousChannelUsername?: string;
       isHidden?: boolean;
       /** @enum {string} */
       visibility: 'public' | 'home' | 'followers' | 'specified';
@@ -4619,6 +4645,8 @@ export type components = {
       pinnedNoteIds: string[];
       color: string;
       isArchived: boolean;
+      /** @enum {string|null} */
+      anonymousStrategy: 'daily' | 'weekly' | 'monthly' | 'yearly' | 'manual';
       usersCount: number;
       notesCount: number;
       isSensitive: boolean;
@@ -4837,7 +4865,7 @@ export type components = {
     RoleCondFormulaValueUserSettingBooleanSchema: {
       id: string;
       /** @enum {string} */
-      type: 'isSuspended' | 'isLocked' | 'isBot' | 'isCat' | 'isExplorable';
+      type: 'isSuspended' | 'isLocked' | 'isBot' | 'isCat' | 'isExplorable' | 'isEmailVerified' | 'isTwoFactorEnabled' | 'isPasswordLessLoginEnabled';
     };
     RoleCondFormulaValueAssignedRole: {
       id: string;
@@ -4917,10 +4945,12 @@ export type components = {
       inviteExpirationTime: number;
       canManageCustomEmojis: boolean;
       canManageAvatarDecorations: boolean;
+      canManageAds: boolean;
       canSearchNotes: boolean;
       canUseTranslator: boolean;
       canHideAds: boolean;
       driveCapacityMb: number;
+      driveUploadBandwidthDurationHrCapacityMbPairs: [number, number][];
       alwaysMarkNsfw: boolean;
       canUpdateBioMedia: boolean;
       pinLimit: number;
@@ -6969,6 +6999,8 @@ export type operations = {
           license?: string | null;
           isSensitive?: boolean;
           localOnly?: boolean;
+          hidden?: boolean;
+          conspicuousScale?: number;
           roleIdsThatCanBeUsedThisEmojiAsReaction?: string[];
         };
       };
@@ -7651,6 +7683,9 @@ export type operations = {
           license?: string | null;
           isSensitive?: boolean;
           localOnly?: boolean;
+          /** @default false */
+          hidden?: boolean;
+          conspicuousScale?: number;
           roleIdsThatCanBeUsedThisEmojiAsReaction?: string[];
         };
       };
@@ -8273,6 +8308,7 @@ export type operations = {
             prohibitedWords: string[];
             prohibitedWordsForNameOfUser: string[];
             bannedEmailDomains?: string[];
+            bannedMediaDomains?: string[];
             preservedUsernames: string[];
             hcaptchaSecretKey: string | null;
             mcaptchaSecretKey: string | null;
@@ -10673,6 +10709,7 @@ export type operations = {
           enableIdenticonGeneration?: boolean;
           serverRules?: string[];
           bannedEmailDomains?: string[];
+          bannedMediaDomains?: string[];
           preservedUsernames?: string[];
           manifestJsonOverride?: string;
           enableFanoutTimeline?: boolean;
@@ -12022,6 +12059,10 @@ export type operations = {
           bannerId?: string | null;
           color?: string;
           isSensitive?: boolean | null;
+          /** @enum {string|null} */
+          anonymousStrategy?: 'daily' | 'weekly' | 'monthly' | 'yearly' | 'manual';
+          /** @default true */
+          requirePublicWriteAccess?: boolean;
           allowRenoteToExternal?: boolean | null;
         };
       };
@@ -12686,6 +12727,9 @@ export type operations = {
           /** Format: misskey:id */
           bannerId?: string | null;
           isArchived?: boolean | null;
+          requirePublicWriteAccess?: boolean | null;
+          /** @default false */
+          regenerateAnonymousSalt?: boolean;
           pinnedNoteIds?: string[];
           color?: string;
           isSensitive?: boolean | null;
@@ -14209,6 +14253,11 @@ export type operations = {
           'application/json': {
             capacity: number;
             usage: number;
+            uploadBandwidths: {
+                duration: number;
+                capacity: number;
+                usage: number;
+              }[];
           };
         };
       };
@@ -18253,6 +18302,7 @@ export type operations = {
         'application/json': {
           password: string;
           token?: string | null;
+          state?: string | null;
           name: string;
           credential: Record<string, never>;
         };
@@ -19846,6 +19896,58 @@ export type operations = {
       };
     };
   };
+  /**
+   * i/jobs
+   * @description No description provided.
+   *
+   * **Internal Endpoint**: This endpoint is an API for the misskey mainframe and is not intended for use by third parties.
+   * **Credential required**: *Yes*
+   */
+  i___jobs: {
+    responses: {
+      /** @description OK (with results) */
+      200: {
+        content: {
+          'application/json': ({
+              id: string;
+              name: string;
+              progress: number | null;
+              timestamp: number;
+            })[];
+        };
+      };
+      /** @description Client error */
+      400: {
+        content: {
+          'application/json': components['schemas']['Error'];
+        };
+      };
+      /** @description Authentication error */
+      401: {
+        content: {
+          'application/json': components['schemas']['Error'];
+        };
+      };
+      /** @description Forbidden error */
+      403: {
+        content: {
+          'application/json': components['schemas']['Error'];
+        };
+      };
+      /** @description I'm Ai */
+      418: {
+        content: {
+          'application/json': components['schemas']['Error'];
+        };
+      };
+      /** @description Internal server error */
+      500: {
+        content: {
+          'application/json': components['schemas']['Error'];
+        };
+      };
+    };
+  };
   /**
    * i/move
    * @description No description provided.
@@ -21028,6 +21130,8 @@ export type operations = {
           followersVisibility?: 'public' | 'followers' | 'private';
           /** Format: misskey:id */
           pinnedPageId?: string | null;
+          /** Format: misskey:id */
+          pinnedGalleryPostId?: string | null;
           mutedWords?: (string[] | string)[];
           hardMutedWords?: (string[] | string)[];
           mutedInstances?: string[];
@@ -22457,6 +22561,8 @@ export type operations = {
           renoteId?: string | null;
           /** Format: misskey:id */
           channelId?: string | null;
+          /** Format: misskey:id */
+          anonymouslySendToUserId?: string | null;
           text?: string | null;
           fileIds?: string[];
           mediaIds?: string[];
@@ -22528,6 +22634,8 @@ export type operations = {
         'application/json': {
           /** Format: misskey:id */
           noteId: string;
+          /** @default false */
+          block?: boolean;
         };
       };
     };
@@ -23427,6 +23535,8 @@ export type operations = {
           sinceId?: string;
           /** Format: misskey:id */
           untilId?: string;
+          sinceDate?: number;
+          untilDate?: number;
           /** @default 10 */
           limit?: number;
           /** @default 0 */
@@ -23443,6 +23553,8 @@ export type operations = {
            * @default null
            */
           channelId?: string | null;
+          /** @default false */
+          internal?: boolean;
         };
       };
     };
@@ -26345,6 +26457,8 @@ export type operations = {
            * @default null
            */
           hostname?: string | null;
+          /** @default false */
+          excludePinned?: boolean;
         };
       };
     };
@@ -26444,6 +26558,80 @@ export type operations = {
       };
     };
   };
+  /**
+   * users/avatars
+   * @description List avatars of users.
+   *
+   * **Credential required**: *No*
+   */
+  users___avatars: {
+    requestBody: {
+      content: {
+        'application/json': {
+          users: ({
+              /** Format: misskey:id */
+              id?: string;
+              username?: string;
+              /** @description The local host is represented with `null`. */
+              host?: string | null;
+            })[];
+        };
+      };
+    };
+    responses: {
+      /** @description OK (with results) */
+      200: {
+        content: {
+          'application/json': ({
+              /** Format: id */
+              id: string;
+              /** @example 藍 */
+              name: string | null;
+              /** @example ai */
+              username: string;
+              /**
+               * @description The local host is represented with `null`.
+               * @example misskey.example.com
+               */
+              host: string | null;
+              /** Format: url */
+              avatarUrl: string;
+              avatarBlurhash: string | null;
+            })[];
+        };
+      };
+      /** @description Client error */
+      400: {
+        content: {
+          'application/json': components['schemas']['Error'];
+        };
+      };
+      /** @description Authentication error */
+      401: {
+        content: {
+          'application/json': components['schemas']['Error'];
+        };
+      };
+      /** @description Forbidden error */
+      403: {
+        content: {
+          'application/json': components['schemas']['Error'];
+        };
+      };
+      /** @description I'm Ai */
+      418: {
+        content: {
+          'application/json': components['schemas']['Error'];
+        };
+      };
+      /** @description Internal server error */
+      500: {
+        content: {
+          'application/json': components['schemas']['Error'];
+        };
+      };
+    };
+  };
   /**
    * users/clips
    * @description Show all clips this user owns.
@@ -27875,6 +28063,8 @@ export type operations = {
         'application/json': {
           /** Format: misskey:id */
           userId: string;
+          /** Format: misskey:id */
+          noteId?: string;
           comment: string;
         };
       };
diff --git a/pnpm-lock.yaml b/pnpm-lock.yaml
index 7524559633..518ac2e748 100644
--- a/pnpm-lock.yaml
+++ b/pnpm-lock.yaml
@@ -317,6 +317,9 @@ importers:
       node-fetch:
         specifier: 3.3.2
         version: 3.3.2
+      node-sql-parser:
+        specifier: 5.3.6
+        version: 5.3.6
       nodemailer:
         specifier: 6.10.0
         version: 6.10.0
@@ -344,6 +347,9 @@ importers:
       pg:
         specifier: 8.13.3
         version: 8.13.3
+      pg-query-stream:
+        specifier: 4.7.3
+        version: 4.7.3(pg@8.13.3)
       pkce-challenge:
         specifier: 4.1.0
         version: 4.1.0
@@ -418,7 +424,7 @@ importers:
         version: 4.2.0
       typeorm:
         specifier: 0.3.20
-        version: 0.3.20(ioredis@5.5.0)(pg@8.13.3)
+        version: 0.3.20(ioredis@5.5.0)(pg-query-stream@4.7.3(pg@8.13.3))(pg@8.13.3)
       typescript:
         specifier: 5.7.3
         version: 5.7.3
@@ -4497,6 +4503,9 @@ packages:
   '@types/offscreencanvas@2019.7.0':
     resolution: {integrity: sha512-PGcyveRIpL1XIqK8eBsmRBt76eFgtzuPiSTyKHZxnGemp2yzGzWpjYKAfK3wIMiU7eH+851yEpiuP8JZerTmWg==}
 
+  '@types/pegjs@0.10.6':
+    resolution: {integrity: sha512-eLYXDbZWXh2uxf+w8sXS8d6KSoXTswfps6fvCUuVAGN8eRpfe7h9eSRydxiSJvo9Bf+GzifsDOr9TMQlmJdmkw==}
+
   '@types/pg-pool@2.0.6':
     resolution: {integrity: sha512-TaAUE5rq2VQYxab5Ts7WZhKNmuN78Q6PiFonTDdpbx8a1H0M1vhy3rhiMjl+e2iHmogyMw7jZF4FrE6eJUy5HQ==}
 
@@ -5220,6 +5229,10 @@ packages:
     resolution: {integrity: sha512-aVNobHnJqLiUelTaHat9DZ1qM2w0C0Eym4LPI/3JxOnSokGVdsl1T1kN7TFvsEAD8G47A6VKQ0TVHqbBnYMJlQ==}
     engines: {node: '>=12.0.0'}
 
+  big-integer@1.6.52:
+    resolution: {integrity: sha512-QxD8cf2eVqJOOz63z6JIN9BzvVs/dlySa5HGSBH5xtR8dPteIRQnBxxKqkNTiT6jbDTF6jAfrd4oMcND9RGbQg==}
+    engines: {node: '>=0.6'}
+
   bin-version-check@5.1.0:
     resolution: {integrity: sha512-bYsvMqJ8yNGILLz1KP9zKLzQ6YpljV3ln1gqhuLkUtyfGi3qXKGuK2p+U4NAvjVFzDFiBBtOpCOSFNuYYEGZ5g==}
     engines: {node: '>=12'}
@@ -8431,6 +8444,10 @@ packages:
   node-releases@2.0.19:
     resolution: {integrity: sha512-xxOWJsBKtzAq7DY0J+DTzuz58K8e7sJbdgwkbMWQe8UYB6ekmsQ45q0M/tJDsGaZmbC+l7n57UV8Hl5tHxO9uw==}
 
+  node-sql-parser@5.3.6:
+    resolution: {integrity: sha512-TXak4rIvmpL7Ap1BJQGfAGewoELQGZ/Rf9YYt9rKz73ykDp3HNnXT08jDFC/8FjqiPoRxcPNVecSwTmjz7LrFA==}
+    engines: {node: '>=8'}
+
   nodemailer@6.10.0:
     resolution: {integrity: sha512-SQ3wZCExjeSatLE/HBaXS5vqUOQk6GtBdIIKxiFdmm01mOQZX/POJkO3SUX1wDiYcwUOJwT23scFSC9fY2H8IA==}
     engines: {node: '>=6.0.0'}
@@ -8771,6 +8788,11 @@ packages:
   pg-connection-string@2.7.0:
     resolution: {integrity: sha512-PI2W9mv53rXJQEOb8xNR8lH7Hr+EKa6oJa38zsK0S/ky2er16ios1wLKhZyxzD7jUReiWokc9WK5nxSnC7W1TA==}
 
+  pg-cursor@2.12.3:
+    resolution: {integrity: sha512-2koS4+f+oCJnJ6pTdUlp3TkwOCjlPr9lmwcTDPeXZv3eiXDCxZeBC7T0+LldA9+mhqdcO8WDqbuFfdFEJm9YLw==}
+    peerDependencies:
+      pg: ^8
+
   pg-int8@1.0.1:
     resolution: {integrity: sha512-WCtabS6t3c8SkpDBUlb1kjOs7l66xsGdKpIPZsg4wR+B3+u9UAum2odSsF9tnvxg80h4ZxLWMy4pRjOsFIqQpw==}
     engines: {node: '>=4.0.0'}
@@ -8790,6 +8812,11 @@ packages:
   pg-protocol@1.7.1:
     resolution: {integrity: sha512-gjTHWGYWsEgy9MsY0Gp6ZJxV24IjDqdpTW7Eh0x+WfJLFsm/TJx1MzL6T0D88mBvkpxotCQ6TwW6N+Kko7lhgQ==}
 
+  pg-query-stream@4.7.3:
+    resolution: {integrity: sha512-WvsdwYXrvIXNZJflX/2MEQLjKs5TopZWeiyam1zcaFfhwQtL19ENOvGdzVsihGsbsNGdVRU5yiqg2G5p06UAbg==}
+    peerDependencies:
+      pg: ^8
+
   pg-types@2.2.0:
     resolution: {integrity: sha512-qTAAlrEsl8s4OiEQY69wDvcMIdQN6wdz5ojQiOy6YRMuynxenON0O5oCpJI6lshc6scgAY8qvJ2On/p+CXY0GA==}
     engines: {node: '>=4'}
@@ -10776,6 +10803,9 @@ packages:
   vue-component-type-helpers@2.2.2:
     resolution: {integrity: sha512-6lLY+n2xz2kCYshl59mL6gy8OUUTmkscmDFMO8i7Lj+QKwgnIFUZmM1i/iTYObtrczZVdw7UakPqDTGwVSGaRg==}
 
+  vue-component-type-helpers@2.2.4:
+    resolution: {integrity: sha512-F66p0XLbAu92BRz6kakHyAcaUSF7HWpWX/THCqL0TxySSj7z/nok5UUMohfNkkCm1pZtawsdzoJ4p1cjNqCx0Q==}
+
   vue-demi@0.14.7:
     resolution: {integrity: sha512-EOG8KXDQNwkJILkx/gPcoL/7vH+hORoBaKgGe+6W7VFMvCYJfmF2dGbvgDroVnI8LU7/kTu8mbjRZGBU1z9NTA==}
     engines: {node: '>=12'}
@@ -14379,7 +14409,7 @@ snapshots:
       ts-dedent: 2.2.0
       type-fest: 2.19.0
       vue: 3.5.13(typescript@5.7.3)
-      vue-component-type-helpers: 2.2.2
+      vue-component-type-helpers: 2.2.4
 
   '@stylistic/eslint-plugin@2.13.0(eslint@9.20.1)(typescript@5.7.3)':
     dependencies:
@@ -14833,6 +14863,8 @@ snapshots:
 
   '@types/offscreencanvas@2019.7.0': {}
 
+  '@types/pegjs@0.10.6': {}
+
   '@types/pg-pool@2.0.6':
     dependencies:
       '@types/pg': 8.11.11
@@ -14846,7 +14878,7 @@ snapshots:
   '@types/pg@8.6.1':
     dependencies:
       '@types/node': 22.13.4
-      pg-protocol: 1.7.0
+      pg-protocol: 1.7.1
       pg-types: 2.2.0
 
   '@types/prop-types@15.7.5': {}
@@ -15738,6 +15770,8 @@ snapshots:
     dependencies:
       open: 8.4.2
 
+  big-integer@1.6.52: {}
+
   bin-version-check@5.1.0:
     dependencies:
       bin-version: 6.0.0
@@ -19802,6 +19836,11 @@ snapshots:
 
   node-releases@2.0.19: {}
 
+  node-sql-parser@5.3.6:
+    dependencies:
+      '@types/pegjs': 0.10.6
+      big-integer: 1.6.52
+
   nodemailer@6.10.0: {}
 
   nodemon@3.1.9:
@@ -20141,6 +20180,10 @@ snapshots:
 
   pg-connection-string@2.7.0: {}
 
+  pg-cursor@2.12.3(pg@8.13.3):
+    dependencies:
+      pg: 8.13.3
+
   pg-int8@1.0.1: {}
 
   pg-numeric@1.0.2: {}
@@ -20153,6 +20196,11 @@ snapshots:
 
   pg-protocol@1.7.1: {}
 
+  pg-query-stream@4.7.3(pg@8.13.3):
+    dependencies:
+      pg: 8.13.3
+      pg-cursor: 2.12.3(pg@8.13.3)
+
   pg-types@2.2.0:
     dependencies:
       pg-int8: 1.0.1
@@ -21934,7 +21982,7 @@ snapshots:
 
   typedarray@0.0.6: {}
 
-  typeorm@0.3.20(ioredis@5.5.0)(pg@8.13.3):
+  typeorm@0.3.20(ioredis@5.5.0)(pg-query-stream@4.7.3(pg@8.13.3))(pg@8.13.3):
     dependencies:
       '@sqltools/formatter': 1.2.5
       app-root-path: 3.1.0
@@ -21954,6 +22002,7 @@ snapshots:
     optionalDependencies:
       ioredis: 5.5.0
       pg: 8.13.3
+      pg-query-stream: 4.7.3(pg@8.13.3)
     transitivePeerDependencies:
       - supports-color
 
@@ -22259,6 +22308,8 @@ snapshots:
 
   vue-component-type-helpers@2.2.2: {}
 
+  vue-component-type-helpers@2.2.4: {}
+
   vue-demi@0.14.7(vue@3.5.13(typescript@5.7.3)):
     dependencies:
       vue: 3.5.13(typescript@5.7.3)
